{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genome plotter\n",
    "\n",
    "This ipython script was written to create a visual representation of the human genome, reflecting its chemical composition (GC content), distribution of protein coding genes, and location of know genome wide association signals.\n",
    "\n",
    "### Requirements:\n",
    "\n",
    "* Previously downloaded and processed genomic data (prepared by `prepare_data.sh`): chunked chromosomes in bed format, gwas catalog, gencode file.\n",
    "* non-standard python libraries: [pandas](http://pandas.pydata.org/), [numpy](http://www.numpy.org/), [cairosvg](http://cairosvg.org/), [pybedtools](https://pythonhosted.org/pybedtools/)\n",
    "\n",
    "\n",
    "### About the data:\n",
    "\n",
    "In this work I was using the GRCH38 build of the human genome (Ensembl release 83), GENCODE version , the GWAS catalog was downloaded on 2015.06.16 (genomic coordinates also in GRCH37), the downloadable GWAS catalog was extended with an in-house maintained positive control collection. \n",
    "\n",
    "\n",
    "### Further readings and references:\n",
    "\n",
    "* Human genome:\n",
    "\n",
    "* GWAS:\n",
    "\n",
    "### Data sources:\n",
    "\n",
    "* [Ensembl](Ensembl.org): the European genome database, where all known genomic information is aggregated including genes, transcripts, variations, regulatory elements and many more. They provide access to the human genome via ftp server.\n",
    "\n",
    "* [GWAS catalog](https://www.ebi.ac.uk/gwas/): manually curated collection of variations in the human genome with known phenotypical changes including [breast size](https://www.ebi.ac.uk/gwas/search?query=Breast%20size) and [polytical ideology](https://www.ebi.ac.uk/gwas/search?query=Political%20ideology). \n",
    "\n",
    "* [GENCODE](http://www.gencodegenes.org/): ultimate resource of annotated genetic elements example genes, transcripts, exons etc. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Last modified: 11/04/2016\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "print \"Last modified:\", (time.strftime(\"%d/%m/%Y\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stage 1.\n",
    "\n",
    "In the first stage we further process the sequence data, and integrate GENCODE and GWAS data into a single dataframe and save it for plotting.\n",
    "\n",
    "### Steps:\n",
    "\n",
    "1. Reading bedfiles with the chunk numbers and the GC content.\n",
    "2. Run bedtools to find out which chunks overlap with protein coding genes and GWAS hits.\n",
    "3. Based on the GC content and the genetic overlapping, a color is assigned to each chunk.\n",
    "4. Constructing a data frame with the above mentioned data.\n",
    "5. Save dataframe in binary file (pickle)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "Importig libraries:\n",
    "'''\n",
    "\n",
    "import gzip # For reading data\n",
    "import pybedtools # For finding overlap between our chunks and genes and GWAS signals\n",
    "import pandas # data handling\n",
    "import numpy as np # Working with large arrays\n",
    "import colorsys # Generate color gradient.\n",
    "import cairosvg # Converting svg to image\n",
    "import pickle # Saving dataframes to disk\n",
    "import os.path # checking if the datafies are already there."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "'''\n",
    "Files used for the anaysis. Let's check if they are really there:\n",
    "'''\n",
    "\n",
    "workingDir = os.getcwd()\n",
    "Chromosome_file_loc = workingDir + \"/data/Processed_chr%s.bed.gz\"\n",
    "GWAS_file_loc = workingDir + \"/data/processed_GWAS.bed.gz\"\n",
    "GENCODE_file_loc = workingDir + \"/data/processed_GENCODE.bed.gz\"\n",
    "outputDir = workingDir + \"/processed_data/dataframe_chr%s.pkl\"\n",
    "\n",
    "if not os.path.isfile(GWAS_file_loc):\n",
    "    print \"GWAS file is missing. Run `prepare_data.sh` first!\"\n",
    "if not os.path.isfile(GENCODE_file_loc):\n",
    "    print \"GENCODE file is missing. Run `prepare_data.sh` first!\"\n",
    "if not os.path.isfile(Chromosome_file_loc % 11):\n",
    "    print \"Processed chromosome files are missing! Run `prepare_data.sh` first!\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "Functions for generating nice colors and gradients\n",
    "'''\n",
    "def linear_gradient(start_hex, finish_hex=\"#FFFFFF\", n=10):\n",
    "    ''' \n",
    "    returns a gradient list of (n) colors between\n",
    "    two hex colors. start_hex and finish_hex\n",
    "    should be the full six-digit color string,\n",
    "    inlcuding the sharp sign (eg \"#FFFFFF\") \n",
    "    '''\n",
    "    # Starting and ending colors in RGB form\n",
    "    s = hex_to_RGB(start_hex)\n",
    "    f = hex_to_RGB(finish_hex)\n",
    "    \n",
    "    # Initilize a list of the output colors with the starting color\n",
    "    RGB_list = [start_hex]\n",
    "    \n",
    "    # Calcuate a color at each evenly spaced value of t from 1 to n\n",
    "    for t in range(1, n):\n",
    "    \n",
    "        # Interpolate RGB vector for color at the current value of t\n",
    "        curr_vector = [\n",
    "            int(s[j] + (float(t)/(n-1))*(f[j]-s[j]))\n",
    "            for j in range(3)\n",
    "        ]\n",
    "\n",
    "        # Add it to our list of output colors\n",
    "        RGB_list.append(RGB_to_hex(curr_vector))\n",
    "\n",
    "    return RGB_list\n",
    "\n",
    "def hex_to_RGB(hex):\n",
    "    ''' \"#FFFFFF\" -> [255,255,255] '''\n",
    "    # Pass 16 to the integer function for change of base\n",
    "    return [int(hex[i:i+2], 16) for i in range(1,6,2)]\n",
    "\n",
    "\n",
    "def RGB_to_hex(RGB):\n",
    "    ''' [255,255,255] -> \"#FFFFFF\" '''\n",
    "    # Components need to be integers for hex to make sense\n",
    "    RGB = [int(x) for x in RGB]\n",
    "    return \"#\"+\"\".join([\"0{0:x}\".format(v) if v < 16 else\n",
    "            \"{0:x}\".format(v) for v in RGB])\n",
    "\n",
    "def get_color(row, colors):\n",
    "    '''\n",
    "    Based on the values in the submitted row, this\n",
    "    function picks a color from the color dictionary, and returns it.\n",
    "    '''\n",
    "    \n",
    "    # Built-in parameters:\n",
    "    threshold = 0.7\n",
    "    max_diff_value = 0.15\n",
    "\n",
    "    # At first we get the index:\n",
    "    try:\n",
    "        index = int(float(row[\"GC_content\"])*20)\n",
    "    except:\n",
    "        index = 0\n",
    "    \n",
    "    # Extracting color based on the gene and the index:\n",
    "    try:\n",
    "        color = colors[row[\"Genes\"]][index]\n",
    "    except:\n",
    "        color = colors[0][index]\n",
    "        \n",
    "    # Ok, we have the color, now based on the column we make it a bit darker:\n",
    "    if row[\"Column_frac\"] > threshold:\n",
    "        diff = (row[\"Column_frac\"] - threshold)/(1 - threshold)\n",
    "        factor = 1 - max_diff_value*diff \n",
    "        \n",
    "        # Get rgb code of the hexa code:\n",
    "        rgb_code = hex_to_RGB(color)\n",
    "        \n",
    "        # Get the hls code of the rgb:\n",
    "        hls_code = colorsys.rgb_to_hls(rgb_code[0]/float(255), \n",
    "                                       rgb_code[1]/float(255), \n",
    "                                       rgb_code[2]/float(255))\n",
    "        \n",
    "        # Get the modifed rgb code:\n",
    "        new_rgb = colorsys.hls_to_rgb(hls_code[0], \n",
    "                                      hls_code[1]*factor, \n",
    "                                      hls_code[2])\n",
    "        \n",
    "        # Get the modifed hexacode:\n",
    "        color = RGB_to_hex([new_rgb[0] * 255,\n",
    "                           new_rgb[1] * 255,\n",
    "                           new_rgb[2] * 255])\n",
    "        \n",
    "    return color  \n",
    "\n",
    "def def_chromosome_colors(chromosome):\n",
    "    '''\n",
    "    This function retruns with an array of colors based on the chromosome number.\n",
    "    input: chromosome number from 0 to 23 \n",
    "    output: [[ten colors for intergenic], [ten colors for genic region]]\n",
    "    \n",
    "    (Ten is the step number of the gradient showing GC content in the given chunk.)\n",
    "    '''\n",
    "\n",
    "    # The colors will be picked from the following array:\n",
    "    color_set = [\n",
    "        [\"#b74242\", \"#b79e42\"],\n",
    "        [\"#b75f42\", \"#b3b742\"],\n",
    "        [\"#b77c42\", \"#96b742\"],\n",
    "        [\"#b79a42\", \"#78b742\"],\n",
    "        [\"#b7b742\", \"#5bb742\"],\n",
    "        [\"#9ab742\", \"#42b746\"],\n",
    "        [\"#7cb742\", \"#42b763\"],\n",
    "        [\"#5fb742\", \"#42b780\"],\n",
    "        [\"#42b742\", \"#42b79e\"],\n",
    "        [\"#42b75f\", \"#42b3b7\"],\n",
    "        [\"#42b77c\", \"#4296b7\"],\n",
    "        [\"#42b79a\", \"#4278b7\"],\n",
    "        [\"#42b7b7\", \"#425bb7\"],\n",
    "        [\"#429ab7\", \"#4642b7\"],\n",
    "        [\"#427cb7\", \"#6342b7\"],\n",
    "        [\"#425fb7\", \"#8042b7\"],\n",
    "        [\"#4242b7\", \"#9e42b7\"],\n",
    "        [\"#5f42b7\", \"#b742b3\"],\n",
    "        [\"#7c42b7\", \"#b74296\"],\n",
    "        [\"#9a42b7\", \"#b74278\"],\n",
    "        [\"#b742b7\", \"#b7425b\"],\n",
    "        [\"#b7429a\", \"#b74642\"],\n",
    "        [\"#b7427c\", \"#b76342\"],\n",
    "        [\"#b7425f\", \"#b78042\"]\n",
    "    ]\n",
    "\n",
    "    # \n",
    "    colors = {}\n",
    "    colors[0] = linear_gradient(color_set[chromosome - 1][0], n=20) # For non genes.\n",
    "    colors[1] = linear_gradient(color_set[chromosome - 1][1], n=20) # For regions overlapping with genes\n",
    "\n",
    "    return colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "'''Function for process data'''\n",
    "\n",
    "def ProcessChromosome(chromosome, Max_rows = 600):\n",
    "    '''\n",
    "    This function reads processed genomic data saved in bed format, comressed with gzip.\n",
    "    \n",
    "    Input is just the chromosome number and a number how many rows we want to \n",
    "    slice the chromosomes (this parameter is optional, default value is 600). \n",
    "    \n",
    "    Output is a dataframe with the following columns:\n",
    "        \"Chunk_no\": integer, \n",
    "        \"GC_content\": float (0..1),\n",
    "        \"Genes\": integer 1 or 0 -> 1 indicates the chunk overlaps with a gene\n",
    "        \"GWAS\": integer 1 or 0 -> 1 indicate overlap with a GWAS signal\n",
    "        \"Colors\": hexadecimal RGB code representing the given chunk.\n",
    "    '''\n",
    "    \n",
    "    # The name of the chromosome is X and Y for number 23 and 24 respectively:\n",
    "    chr_name = chromosome\n",
    "    if chr_name == 23: chr_name = \"X\"\n",
    "    if chr_name == 24: chr_name = \"Y\"\n",
    "\n",
    "    # Importing global variables:\n",
    "    global Chromosome_file_loc\n",
    "    global GWAS_file_loc\n",
    "    global GENCODE_file_loc\n",
    "\n",
    "    # Opening bedfile, read GC contents and chunk number:\n",
    "    chunk_no = []\n",
    "    CG_content = []\n",
    "    with gzip.open(Chromosome_file_loc % chr_name,'r') as bedfile:\n",
    "        for line in bedfile:\n",
    "            line = line.strip()\n",
    "            (chrom, start, end, GC, chunk) = line.split(\"\\t\")\n",
    "\n",
    "            chunk_no.append(chunk)\n",
    "            CG_content.append(GC)\n",
    "            \n",
    "    # As soon as the GC contents are read, we check \n",
    "    # if the chunks are overlapping with genes or GWAS hits:\n",
    "\n",
    "    # Run intersectBed query:\n",
    "    data_file = pybedtools.BedTool((Chromosome_file_loc % chr_name))\n",
    "    GENCODE_file = pybedtools.BedTool(GENCODE_file_loc)\n",
    "    GWAS_file = pybedtools.BedTool(GWAS_file_loc)\n",
    "\n",
    "    # Get intersecting genes:\n",
    "    GencodeIntersect = data_file.intersect(GENCODE_file, wa = True)\n",
    "\n",
    "    # Get intersecting GWAS signals:\n",
    "    GWASIntersect = data_file.intersect(GWAS_file, wa = True)\n",
    "    \n",
    "    # Now we have to loop through all intersecting values:\n",
    "    Genes = np.zeros(len(CG_content), dtype=np.int)\n",
    "    GWAS = np.zeros(len(CG_content), dtype=np.int)\n",
    "    chunks_in_genes = [hit.fields[4] for hit in GencodeIntersect]\n",
    "    chunks_in_gwas  = [hit.fields[4] for hit in GWASIntersect]\n",
    "    \n",
    "    # elements of the arrays corresponding to overlapping chunks\n",
    "    # will be updated to 1 from 0:\n",
    "    for index in chunks_in_genes:\n",
    "        Genes[int(index)] = 1\n",
    "\n",
    "    for index in chunks_in_gwas:\n",
    "        GWAS[int(index)] = 1\n",
    "\n",
    "    # Data organized into dataframe:\n",
    "    df = pandas.DataFrame({\n",
    "            \"Chunk_no\": chunk_no,\n",
    "            \"GC_content\": CG_content,\n",
    "            \"Genes\": Genes,\n",
    "            \"GWAS\": GWAS\n",
    "        })    \n",
    "        \n",
    "    return df\n",
    "\n",
    "\n",
    "def color_chromosome(df, chromosome, Max_rows = 600):\n",
    "    '''\n",
    "    Once we have the dataframe of the chunks we can assign colors based on the \n",
    "    chromosome number, CG content, gene annotation.\n",
    "    '''\n",
    "    # Based on the maximum number of rows, we assign a row location \n",
    "    # for each element:\n",
    "    # Number of lines the whole genome is broken up\n",
    "    Max_columns = len(df) / Max_rows\n",
    "\n",
    "    # Calculating the coordinates for each chunk in the final plot, and add to the df:\n",
    "    rows = [int(x)/Max_columns for x in df.Chunk_no.tolist()]\n",
    "    column = [int(x) - Max_columns*(int(x)/Max_columns) for x in df.Chunk_no.tolist()]\n",
    "    df[\"Row_no\"] = pandas.Series(rows)\n",
    "    df[\"Column_no\"] = pandas.Series(column)\n",
    "    df[\"Column_frac\"] = [float(x) / Max_columns for x in column]\n",
    "\n",
    "    # Picking colors for the chromosome\n",
    "    colors = def_chromosome_colors(chromosome)\n",
    "\n",
    "    # Assigning colors from the above defined dictionary:\n",
    "    df[\"color\"] = df.apply(get_color, axis=1, args=(colors,))    \n",
    "\n",
    "    return df\n",
    "\n",
    "def plot_chromosome(df, chromosome, filename, rows = 600):\n",
    "    '''\n",
    "    Genomic data processed, colors have been assigned to each chunk.\n",
    "    Now create an svg image, then render it in a png.\n",
    "    '''\n",
    "    width = int(len(df)/rows)\n",
    "    chromosome_name = chromosome\n",
    "    # Although I decided to focus on the autosomal chromosomes, the script is ready to deal with\n",
    "    # the sex chromosomes.\n",
    "    if chromosome_name == 23: chromosome_name = \"X\"\n",
    "    if chromosome_name == 24: chromosome_name = \"Y\"        \n",
    "\n",
    "    # The size of box which represent a single chunk of the genome:\n",
    "    box_width = 2\n",
    "    box_height = 2\n",
    "\n",
    "    # svg file is saved:\n",
    "    #f = open('chromosome1.html', 'w')\n",
    "    plot = '<svg width=\"%s\" height=\"%s\">\\n' % (width*1.5*box_width, rows*1.5*box_height)\n",
    "\n",
    "    # Plotting genome based on GC content and gene annotation:\n",
    "    for index, color in enumerate(df.color):\n",
    "\n",
    "        # Get x,y coordinates based on index:\n",
    "        y = (index / width) * (box_height + 1)\n",
    "        x = (index - (index / width)*width)*(box_width+1)\n",
    "\n",
    "        # Generate svg line:\n",
    "        plot +='<rect x=\"%s\" y=\"%s\" width=\"%s\" height=\"%s\" style=\"stroke-width:1;stroke:%s; fill: %s\" />\\n' % (x, y, box_width + 1 , box_height + 1, color, color)\n",
    "\n",
    "    # looking through gwas annotation:\n",
    "    for index, gwas in enumerate(df.GWAS):\n",
    "        if gwas == 1:\n",
    "            # based on the index of the given line, we get the x,y coordinates:\n",
    "            y = (index / width) * (box_height + 1) + 1.5\n",
    "            x = (index - (index / width)*width)*(box_width+1) + 1.5\n",
    "\n",
    "            # Drawing circle around the defined point:\n",
    "            plot += '<circle cx=\"%s\" cy=\"%s\" r=\"3\" stroke=\"%s\" stroke-width=\"1\" fill=\"%s\" />\\n' % (x, y, \"black\", \"black\")\n",
    "\n",
    "    ###\n",
    "    ### Drawing legend for the figure.\n",
    "    ### A box with the name of the chromosome explanation of the colors\n",
    "    ### Scale bars etc.\n",
    "    ###\n",
    "\n",
    "    # Adding a bigger rectangle, for legend:\n",
    "    legend_x = 15\n",
    "    legend_y = 15\n",
    "    legend_width = 250\n",
    "    legend_height = 220\n",
    "    plot += '<g>\\n'\n",
    "    plot += '<rect x=\"%s\" y=\"%s\" rx=\"20\" ry=\"20\" width=\"%s\" height=\"%s\" style=\"fill:white;stroke:black;stroke-width:2\" />\\n' % (\n",
    "            legend_x, legend_y, legend_width, legend_height)\n",
    "\n",
    "    # Adding chromosome number:\n",
    "    plot += '<text x=\"140\" y=\"46\" text-anchor=\"middle\" font-family=\"Verdana\" font-size=\"30\">chr%s</text>\\n' % (chromosome_name)\n",
    "\n",
    "    # Adding text: (GC content)\n",
    "    plot += '<text x=\"95\" y=\"80\" text-anchor=\"middle\" font-family=\"Verdana\" font-size=\"13\">GC content</text>\\n'\n",
    "\n",
    "    # For the color section of the legend, we have to generate a color gradient:\n",
    "    colors = def_chromosome_colors(chromosome)\n",
    "\n",
    "    # Adding gene colors:\n",
    "    y = 85\n",
    "    x = 20\n",
    "    height = 15\n",
    "    width = 7\n",
    "    for index, col in enumerate (reversed(colors[1])):\n",
    "        x_coord = x+width*int(index)\n",
    "        plot += '<rect x=\"%s\" y=\"%s\" width=\"%s\" height=\"%s\" style=\"stroke-width:1;stroke:%s; fill: %s\" />\\n' %(x_coord, y, width, height, col, col)\n",
    "    plot += '<text x=\"%s\" y=\"%s\" font-family=\"Verdana\" font-size=\"17\">%s</text>\\n' %(x_coord + 15, y + 12, \"Genes\")\n",
    "\n",
    "    # Adding background colors:\n",
    "    y = 105\n",
    "    for index, col in enumerate (reversed(colors[0])):\n",
    "        x_coord = x+width*int(index)\n",
    "        plot += '<rect x=\"%s\" y=\"%s\" width=\"%s\" height=\"%s\" style=\"stroke-width:1;stroke:%s; fill: %s\" />\\n' %(x_coord, y, width, height, col, col)\n",
    "    plot += '<text x=\"%s\" y=\"%s\" font-family=\"Verdana\" font-size=\"17\">%s</text>\\n' %(x_coord + 15, y + 12, \"Intergenic\")\n",
    "\n",
    "    # Adding scale bars:\n",
    "    length = 63\n",
    "    x = 40\n",
    "    y = 160\n",
    "    tick = 5\n",
    "\n",
    "    plot += '<text x=\"%s\" y=\"%s\" font-family=\"Verdana\" text-anchor=\"middle\"  font-size=\"15\">15kbp</text>\\n' % (length/2 + x, y - 10)\n",
    "    plot += '<line x1=\"%s\" y1=\"%s\" x2=\"%s\" y2=\"%s\" style=\"stroke:black;stroke-width:3\" />\\n' %(x,y,x + length,y)\n",
    "    plot += '<line x1=\"%s\" y1=\"%s\" x2=\"%s\" y2=\"%s\" style=\"stroke:black;stroke-width:3\" />\\n' %(x,y-tick,x,y+tick)\n",
    "    plot += '<line x1=\"%s\" y1=\"%s\" x2=\"%s\" y2=\"%s\" style=\"stroke:black;stroke-width:3\" />\\n' %(x + length,y-tick,x + length,y+tick)\n",
    "\n",
    "    # Adding verical units: 42 units for 500kbp\n",
    "    # Get the length of one row: \n",
    "    length_row = len(df[df.Row_no == 3]) * 480\n",
    "    length_bar = 50\n",
    "    bar_size_genome = round(length_bar / 2 * length_row/1e6, 2)\n",
    "\n",
    "    x = 140\n",
    "    y = 160\n",
    "    plot += '<text x=\"%s\" y=\"%s\" font-family=\"Verdana\" font-size=\"15\">%sMbp</text>\\n' % (\n",
    "        x + 15, y, bar_size_genome)\n",
    "    plot += '<line x1=\"%s\" y1=\"%s\" x2=\"%s\" y2=\"%s\" style=\"stroke:black;stroke-width:3\" />\\n' % (\n",
    "        x, y-length_bar/2, x, y + length_bar/2)\n",
    "    plot += '<line x1=\"%s\" y1=\"%s\" x2=\"%s\" y2=\"%s\" style=\"stroke:black;stroke-width:3\" />\\n' % (\n",
    "        x-tick,y-length_bar/2,x+tick,y-length_bar/2)\n",
    "    plot += '<line x1=\"%s\" y1=\"%s\" x2=\"%s\" y2=\"%s\" style=\"stroke:black;stroke-width:3\" />\\n' % (\n",
    "        x-tick,y + length_bar/2,x+tick,y + length_bar/2)\n",
    "\n",
    "    # Adding a circle to the bottom of the legend:\n",
    "    circle_r = 5\n",
    "    circle_x = 40 \n",
    "    circle_y = 210\n",
    "    plot += '<circle cx=\"%s\" cy=\"%s\" r=\"%s\" stroke=\"black\" stroke-width=\"1\" fill=\"black\" />\\n' % (\n",
    "            circle_x, circle_y, circle_r)\n",
    "    plot += '<text x=\"%s\" y=\"%s\" font-family=\"Verdana\" font-size=\"15\">%s</text>\\n' % (\n",
    "            circle_x + 20, circle_y + 5, \"Known GWAS signals\")\n",
    "\n",
    "    # Closing legend\n",
    "    plot += '</g>'\n",
    "\n",
    "    #f.write('</svg>\\n</body>\\n</html>\\n')\n",
    "    plot += '</svg>\\n'\n",
    "    #f.close()\n",
    "\n",
    "    fout = open('%s_chr%s.png' % (filename, chromosome),'w')\n",
    "    cairosvg.svg2png(bytestring=plot,write_to=fout)\n",
    "    fout.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing chromosome 1\n",
      "Processing chromosome 2\n",
      "Processing chromosome 3\n",
      "Processing chromosome 4\n",
      "Processing chromosome 5\n",
      "Processing chromosome 6\n",
      "Processing chromosome 7\n",
      "Processing chromosome 8\n",
      "Processing chromosome 9\n",
      "Processing chromosome 10\n",
      "Processing chromosome 11\n",
      "Processing chromosome 12\n",
      "Processing chromosome 13\n",
      "Processing chromosome 14\n",
      "Processing chromosome 15\n",
      "Processing chromosome 16\n",
      "Processing chromosome 17\n",
      "Processing chromosome 18\n",
      "Processing chromosome 19\n",
      "Processing chromosome 20\n",
      "Processing chromosome 21\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Now we loop through all autosomes and generate a plot.\n",
    "In the meanwhile, the dataframes are also saved, and upon a future step,\n",
    "they can be read back withouth calculating them from the scratch.\n",
    "'''\n",
    "for chromosome in range(1, 22):\n",
    "    print \"Processing chromosome %s\" % (chromosome)\n",
    "\n",
    "    # Process chromosome data, assign colors, combine into dataframe:\n",
    "    df = ProcessChromosome(chromosome)\n",
    "\n",
    "    # Pickling dataframes:\n",
    "    filename = outputDir % chromosome\n",
    "    output = open(filename, 'wb')\n",
    "    pickle.dump(df, output)\n",
    "    \n",
    "    # If we jus want ot read df from file:\n",
    "\n",
    "    # Add colors to datafrome:\n",
    "    df = color_chromosome(df, chromosome)\n",
    "    plot_chromosome(df, chromosome, \"plot\", 600)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
