{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genome plotter\n",
    "\n",
    "This ipython script was written to create a visual representation of the human genome, reflecting its chemical composition (GC content), distribution of protein coding genes, and location of know genome wide association signals.\n",
    "\n",
    "### Requirements:\n",
    "\n",
    "* Previously downloaded and processed genomic data (prepared by `prepare_data.sh`): chunked chromosomes in bed format, gwas catalog, gencode file.\n",
    "* non-standard python libraries: [pandas](http://pandas.pydata.org/), [numpy](http://www.numpy.org/), [cairosvg](http://cairosvg.org/), [pybedtools](https://pythonhosted.org/pybedtools/)\n",
    "* Just for the record: once cairo has been installed on an OSX using homebrew, python by default won't be able to link to the libraries. Therefore this path pointing the `cairo/lib` directory has to be exported:\n",
    "\n",
    "```bash\n",
    "export DYLD_FALLBACK_LIBRARY_PATH=~/homebrew/Cellar/cairo/1.14.12/lib/\n",
    "```\n",
    "\n",
    "\n",
    "### About the data:\n",
    "\n",
    "In this work I was using the GRCH38 build of the human genome (Ensembl release 83), GENCODE version , the GWAS catalog was downloaded on 2015.06.16 (genomic coordinates also in GRCH37), the downloadable GWAS catalog was extended with an in-house maintained positive control collection. \n",
    "\n",
    "\n",
    "### Further readings and references:\n",
    "\n",
    "* Human genome:\n",
    "\n",
    "* GWAS:\n",
    "\n",
    "### Data sources:\n",
    "\n",
    "* [Ensembl](Ensembl.org): the European genome database, where all known genomic information is aggregated including genes, transcripts, variations, regulatory elements and many more. They provide access to the human genome via ftp server.\n",
    "\n",
    "* [GWAS catalog](https://www.ebi.ac.uk/gwas/): manually curated collection of variations in the human genome with known phenotypical changes including [breast size](https://www.ebi.ac.uk/gwas/search?query=Breast%20size) and [polytical ideology](https://www.ebi.ac.uk/gwas/search?query=Political%20ideology). \n",
    "\n",
    "* [GENCODE](http://www.gencodegenes.org/): ultimate resource of annotated genetic elements example genes, transcripts, exons etc. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Last modified: 04/01/2018\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "print (\"Last modified:\", (time.strftime(\"%d/%m/%Y\")))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stage 1.\n",
    "\n",
    "In the first stage we further process the sequence data, and integrate GENCODE and GWAS data into a single dataframe and save it for plotting.\n",
    "\n",
    "### Steps:\n",
    "\n",
    "1. Reading bedfiles with the chunk numbers and the GC content.\n",
    "2. Run bedtools to find out which chunks overlap with protein coding genes and GWAS hits.\n",
    "3. Based on the GC content and the genetic overlapping, a color is assigned to each chunk.\n",
    "4. Constructing a data frame with the above mentioned data.\n",
    "5. Save dataframe in binary file (pickle)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "'''\n",
    "Importig libraries:\n",
    "'''\n",
    "\n",
    "import gzip # For reading data\n",
    "import pybedtools # For finding overlap between our chunks and genes and GWAS signals\n",
    "import pandas # data handling\n",
    "import numpy as np # Working with large arrays\n",
    "import colorsys # Generate color gradient.\n",
    "import cairosvg # Converting svg to image\n",
    "import pickle # Saving dataframes to disk\n",
    "import os.path # checking if the datafies are already there."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "'''\n",
    "Files used for the anaysis. Let's check if they are really there:\n",
    "'''\n",
    "\n",
    "workingDir = os.getcwd()\n",
    "Chromosome_file_loc = workingDir + \"/data/Processed_chr%s.bed.gz\"\n",
    "GWAS_file_loc = workingDir + \"/data/processed_GWAS.bed.gz\"\n",
    "GENCODE_file_loc = workingDir + \"/data/GENCODE.merged.bed.gz\"\n",
    "outputDir = workingDir + \"/processed_data/dataframe_chr%s.pkl\"\n",
    "\n",
    "if not os.path.isfile(GWAS_file_loc):\n",
    "    print (\"GWAS file is missing. Run `prepare_data.sh` first!\")\n",
    "if not os.path.isfile(GENCODE_file_loc):\n",
    "    print (\"GENCODE file is missing. Run `prepare_data.sh` first!\")\n",
    "if not os.path.isfile(Chromosome_file_loc % 11):\n",
    "    print (\"Processed chromosome files are missing! Run `prepare_data.sh` first!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "Functions for generating nice colors and gradients\n",
    "'''\n",
    "def linear_gradient(start_hex, finish_hex=\"#FFFFFF\", n=10):\n",
    "    ''' \n",
    "    returns a gradient list of (n) colors between\n",
    "    two hex colors. start_hex and finish_hex\n",
    "    should be the full six-digit color string,\n",
    "    inlcuding the sharp sign (eg \"#FFFFFF\") \n",
    "    '''\n",
    "    # Starting and ending colors in RGB form\n",
    "    s = hex_to_RGB(start_hex)\n",
    "    f = hex_to_RGB(finish_hex)\n",
    "    \n",
    "    # Initilize a list of the output colors with the starting color\n",
    "    RGB_list = [start_hex]\n",
    "    \n",
    "    # Calcuate a color at each evenly spaced value of t from 1 to n\n",
    "    for t in range(1, n):\n",
    "    \n",
    "        # Interpolate RGB vector for color at the current value of t\n",
    "        curr_vector = [\n",
    "            int(s[j] + (float(t)/(n-1))*(f[j]-s[j]))\n",
    "            for j in range(3)\n",
    "        ]\n",
    "\n",
    "        # Add it to our list of output colors\n",
    "        RGB_list.append(RGB_to_hex(curr_vector))\n",
    "\n",
    "    return RGB_list\n",
    "\n",
    "def hex_to_RGB(hex):\n",
    "    ''' \"#FFFFFF\" -> [255,255,255] '''\n",
    "    # Pass 16 to the integer function for change of base\n",
    "    return [int(hex[i:i+2], 16) for i in range(1,6,2)]\n",
    "\n",
    "\n",
    "def RGB_to_hex(RGB):\n",
    "    ''' [255,255,255] -> \"#FFFFFF\" '''\n",
    "    # Components need to be integers for hex to make sense\n",
    "    RGB = [int(x) for x in RGB]\n",
    "    return \"#\"+\"\".join([\"0{0:x}\".format(v) if v < 16 else\n",
    "            \"{0:x}\".format(v) for v in RGB])\n",
    "\n",
    "def get_color(row, colors):\n",
    "    '''\n",
    "    Based on the values in the submitted row, this\n",
    "    function picks a color from the color dictionary, and returns it.\n",
    "    '''\n",
    "    \n",
    "    # Built-in parameters:\n",
    "    threshold = 0.7 \n",
    "    max_diff_value = 0.15\n",
    "\n",
    "    # At first we get the index:\n",
    "    try:\n",
    "        index = int(float(row[\"GC_content\"])*20)\n",
    "    except:\n",
    "        index = 0\n",
    "    \n",
    "    # Extracting color based on the gene and the index:\n",
    "    try:\n",
    "        color = colors[row[\"Genes\"]][index]\n",
    "    except:\n",
    "        color = colors[0][index]\n",
    "        \n",
    "    # Ok, we have the color, now based on the column we make it a bit darker:\n",
    "    if row[\"Column_frac\"] > threshold:\n",
    "        diff = (row[\"Column_frac\"] - threshold)/(1 - threshold)\n",
    "        factor = 1 - max_diff_value*diff \n",
    "        \n",
    "        # Get rgb code of the hexa code:\n",
    "        rgb_code = hex_to_RGB(color)\n",
    "        \n",
    "        # Get the hls code of the rgb:\n",
    "        hls_code = colorsys.rgb_to_hls(rgb_code[0]/float(255), \n",
    "                                       rgb_code[1]/float(255), \n",
    "                                       rgb_code[2]/float(255))\n",
    "        \n",
    "        # Get the modifed rgb code:\n",
    "        new_rgb = colorsys.hls_to_rgb(hls_code[0], \n",
    "                                      hls_code[1]*factor, \n",
    "                                      hls_code[2])\n",
    "        \n",
    "        # Get the modifed hexacode:\n",
    "        color = RGB_to_hex([new_rgb[0] * 255,\n",
    "                           new_rgb[1] * 255,\n",
    "                           new_rgb[2] * 255])\n",
    "        \n",
    "    return color  \n",
    "\n",
    "def def_chromosome_colors(chromosome):\n",
    "    '''\n",
    "    This function retruns with an array of colors based on the chromosome number.\n",
    "    input: chromosome number from 0 to 23 \n",
    "    output: [[ten colors for intergenic], [ten colors for genic region]]\n",
    "    \n",
    "    (Ten is the step number of the gradient showing GC content in the given chunk.)\n",
    "    '''\n",
    "\n",
    "    # The colors will be picked from the following array:\n",
    "    color_set = [\n",
    "        [\"#b74242\", \"#b79e42\"],\n",
    "        [\"#b75f42\", \"#b3b742\"],\n",
    "        [\"#b77c42\", \"#96b742\"],\n",
    "        [\"#b79a42\", \"#78b742\"],\n",
    "        [\"#b7b742\", \"#5bb742\"],\n",
    "        [\"#9ab742\", \"#42b746\"],\n",
    "        [\"#7cb742\", \"#42b763\"],\n",
    "        [\"#5fb742\", \"#42b780\"],\n",
    "        [\"#42b742\", \"#42b79e\"],\n",
    "        [\"#42b75f\", \"#42b3b7\"],\n",
    "        [\"#42b77c\", \"#4296b7\"],\n",
    "        [\"#42b79a\", \"#4278b7\"],\n",
    "        [\"#42b7b7\", \"#425bb7\"],\n",
    "        [\"#429ab7\", \"#4642b7\"],\n",
    "        [\"#427cb7\", \"#6342b7\"],\n",
    "        [\"#425fb7\", \"#8042b7\"],\n",
    "        [\"#4242b7\", \"#9e42b7\"],\n",
    "        [\"#5f42b7\", \"#b742b3\"],\n",
    "        [\"#7c42b7\", \"#b74296\"],\n",
    "        [\"#9a42b7\", \"#b74278\"],\n",
    "        [\"#b742b7\", \"#b7425b\"],\n",
    "        [\"#b7429a\", \"#b74642\"],\n",
    "        [\"#b7427c\", \"#b76342\"],\n",
    "        [\"#b7425f\", \"#b78042\"]\n",
    "    ]\n",
    "\n",
    "    # \n",
    "    colors = {}\n",
    "    colors[0] = linear_gradient(color_set[chromosome - 1][0], n=20) # For non genes.\n",
    "    colors[1] = linear_gradient(color_set[chromosome - 1][1], n=20) # For regions overlapping with genes\n",
    "\n",
    "    return colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "'''Function for process data'''\n",
    "\n",
    "def ProcessChromosome(chromosome, Max_rows = 600):\n",
    "    '''\n",
    "    This function reads processed genomic data saved in bed format, compressed with gzip.\n",
    "    \n",
    "    Input is just the chromosome number and a number how many rows we want to \n",
    "    slice the chromosomes (this parameter is optional, default value is 600). \n",
    "    \n",
    "    Output is a dataframe with the following columns:\n",
    "        \"Chunk_no\": integer, \n",
    "        \"GC_content\": float (0..1),\n",
    "        \"Genes\": integer 1 or 0 -> 1 indicates the chunk overlaps with a gene\n",
    "        \"GWAS\": integer 1 or 0 -> 1 indicate overlap with a GWAS signal\n",
    "        \"Colors\": hexadecimal RGB code representing the given chunk.\n",
    "    '''\n",
    "    \n",
    "    # The name of the chromosome is X and Y for number 23 and 24 respectively:\n",
    "    chr_name = chromosome\n",
    "    if chr_name == 23: chr_name = \"X\"\n",
    "    if chr_name == 24: chr_name = \"Y\"\n",
    "\n",
    "    # Importing global variables:\n",
    "    global Chromosome_file_loc\n",
    "    global GWAS_file_loc\n",
    "    global GENCODE_file_loc\n",
    "\n",
    "    # Opening bedfile, read GC contents and chunk number:\n",
    "    chunk_no = []\n",
    "    CG_content = []\n",
    "    with gzip.open(Chromosome_file_loc % chr_name,'r') as bedfile:\n",
    "        for line in bedfile:\n",
    "            line = line.strip()\n",
    "            (chrom, start, end, GC, chunk) = line.split(\"\\t\")\n",
    "\n",
    "            chunk_no.append(chunk)\n",
    "            CG_content.append(GC)\n",
    "            \n",
    "    # As soon as the GC contents are read, we check \n",
    "    # if the chunks are overlapping with genes or GWAS hits:\n",
    "\n",
    "    # Run intersectBed query:\n",
    "    data_file = pybedtools.BedTool((Chromosome_file_loc % chr_name))\n",
    "    GENCODE_file = pybedtools.BedTool(GENCODE_file_loc)\n",
    "    GWAS_file = pybedtools.BedTool(GWAS_file_loc)\n",
    "\n",
    "    # Get intersecting genes:\n",
    "    GencodeIntersect = data_file.intersect(GENCODE_file, wa = True)\n",
    "\n",
    "    # Get intersecting GWAS signals:\n",
    "    GWASIntersect = data_file.intersect(GWAS_file, wa = True)\n",
    "    \n",
    "    # Now we have to loop through all intersecting values:\n",
    "    Genes = np.zeros(len(CG_content), dtype=np.int)\n",
    "    GWAS = np.zeros(len(CG_content), dtype=np.int)\n",
    "    chunks_in_genes = [hit.fields[4] for hit in GencodeIntersect]\n",
    "    chunks_in_gwas  = [hit.fields[4] for hit in GWASIntersect]\n",
    "    \n",
    "    # elements of the arrays corresponding to overlapping chunks\n",
    "    # will be updated to 1 from 0:\n",
    "    for index in chunks_in_genes:\n",
    "        Genes[int(index)] = 1\n",
    "\n",
    "    for index in chunks_in_gwas:\n",
    "        GWAS[int(index)] = 1\n",
    "\n",
    "    # Data organized into dataframe:\n",
    "    df = pandas.DataFrame({\n",
    "            \"Chunk_no\": chunk_no,\n",
    "            \"GC_content\": CG_content,\n",
    "            \"Genes\": Genes,\n",
    "            \"GWAS\": GWAS\n",
    "        })    \n",
    "        \n",
    "    return df\n",
    "\n",
    "\n",
    "def color_chromosome(df, chromosome, Max_rows = 600):\n",
    "    '''\n",
    "    Once we have the dataframe of the chunks we can assign colors based on the \n",
    "    chromosome number, CG content, gene annotation.\n",
    "    '''\n",
    "    # Based on the maximum number of rows, we assign a row location \n",
    "    # for each element:\n",
    "    # Number of lines the whole genome is broken up\n",
    "    Max_columns = len(df) / Max_rows\n",
    "\n",
    "    # Calculating the coordinates for each chunk in the final plot, and add to the df:\n",
    "    rows = [int(x)/Max_columns for x in df.Chunk_no.tolist()]\n",
    "    column = [int(x) - Max_columns*(int(x)/Max_columns) for x in df.Chunk_no.tolist()]\n",
    "    df[\"Row_no\"] = pandas.Series(rows)\n",
    "    df[\"Column_no\"] = pandas.Series(column)\n",
    "    df[\"Column_frac\"] = [float(x) / Max_columns for x in column]\n",
    "\n",
    "    # Picking colors for the chromosome\n",
    "    colors = def_chromosome_colors(chromosome)\n",
    "\n",
    "    # Assigning colors from the above defined dictionary:\n",
    "    df[\"color\"] = df.apply(get_color, axis=1, args=(colors,))    \n",
    "\n",
    "    return df\n",
    "\n",
    "def plot_chromosome(df, chromosome, filename, rows = 600):\n",
    "    '''\n",
    "    Genomic data processed, colors have been assigned to each chunk.\n",
    "    Now create an svg image, then render it in a png.\n",
    "    '''\n",
    "    width = int(len(df)/rows)\n",
    "    chromosome_name = chromosome\n",
    "    # Although I decided to focus on the autosomal chromosomes, the script is ready to deal with\n",
    "    # the sex chromosomes.\n",
    "    if chromosome_name == 23: chromosome_name = \"X\"\n",
    "    if chromosome_name == 24: chromosome_name = \"Y\"        \n",
    "\n",
    "    # The size of box which represent a single chunk of the genome:\n",
    "    box_width = 2\n",
    "    box_height = 2\n",
    "\n",
    "    # svg file is saved:\n",
    "    #f = open('chromosome1.html', 'w')\n",
    "    plot = '<svg width=\"%s\" height=\"%s\">\\n' % (width*1.5*box_width, rows*1.5*box_height)\n",
    "\n",
    "    # Plotting genome based on GC content and gene annotation:\n",
    "    for index, color in enumerate(df.color):\n",
    "\n",
    "        # Get x,y coordinates based on index:\n",
    "        y = (index / width) * (box_height + 1)\n",
    "        x = (index - (index / width)*width)*(box_width+1)\n",
    "\n",
    "        # Generate svg line:\n",
    "        plot +='<rect x=\"%s\" y=\"%s\" width=\"%s\" height=\"%s\" style=\"stroke-width:1;stroke:%s; fill: %s\" />\\n' % (x, y, box_width + 1 , box_height + 1, color, color)\n",
    "\n",
    "    # looking through gwas annotation:\n",
    "    for index, gwas in enumerate(df.GWAS):\n",
    "        if gwas == 1:\n",
    "            # based on the index of the given line, we get the x,y coordinates:\n",
    "            y = (index / width) * (box_height + 1) + 1.5\n",
    "            x = (index - (index / width)*width)*(box_width+1) + 1.5\n",
    "\n",
    "            # Drawing circle around the defined point:\n",
    "            plot += '<circle cx=\"%s\" cy=\"%s\" r=\"3\" stroke=\"%s\" stroke-width=\"1\" fill=\"%s\" />\\n' % (x, y, \"black\", \"black\")\n",
    "\n",
    "    ###\n",
    "    ### Drawing legend for the figure.\n",
    "    ### A box with the name of the chromosome explanation of the colors\n",
    "    ### Scale bars etc.\n",
    "    ###\n",
    "\n",
    "    # Adding a bigger rectangle, for legend:\n",
    "    legend_x = 15\n",
    "    legend_y = 15\n",
    "    legend_width = 250\n",
    "    legend_height = 220\n",
    "    plot += '<g>\\n'\n",
    "    plot += '<rect x=\"%s\" y=\"%s\" rx=\"20\" ry=\"20\" width=\"%s\" height=\"%s\" style=\"fill:white;stroke:black;stroke-width:2\" />\\n' % (\n",
    "            legend_x, legend_y, legend_width, legend_height)\n",
    "\n",
    "    # Adding chromosome number:\n",
    "    plot += '<text x=\"140\" y=\"46\" text-anchor=\"middle\" font-family=\"Verdana\" font-size=\"30\">chr%s</text>\\n' % (chromosome_name)\n",
    "\n",
    "    # Adding text: (GC content)\n",
    "    plot += '<text x=\"95\" y=\"80\" text-anchor=\"middle\" font-family=\"Verdana\" font-size=\"13\">GC content</text>\\n'\n",
    "\n",
    "    # For the color section of the legend, we have to generate a color gradient:\n",
    "    colors = def_chromosome_colors(chromosome)\n",
    "\n",
    "    # Adding gene colors:\n",
    "    y = 85\n",
    "    x = 20\n",
    "    height = 15\n",
    "    width = 7\n",
    "    for index, col in enumerate (reversed(colors[1])):\n",
    "        x_coord = x+width*int(index)\n",
    "        plot += '<rect x=\"%s\" y=\"%s\" width=\"%s\" height=\"%s\" style=\"stroke-width:1;stroke:%s; fill: %s\" />\\n' %(x_coord, y, width, height, col, col)\n",
    "    plot += '<text x=\"%s\" y=\"%s\" font-family=\"Verdana\" font-size=\"17\">%s</text>\\n' %(x_coord + 15, y + 12, \"Genes\")\n",
    "\n",
    "    # Adding background colors:\n",
    "    y = 105\n",
    "    for index, col in enumerate (reversed(colors[0])):\n",
    "        x_coord = x+width*int(index)\n",
    "        plot += '<rect x=\"%s\" y=\"%s\" width=\"%s\" height=\"%s\" style=\"stroke-width:1;stroke:%s; fill: %s\" />\\n' %(x_coord, y, width, height, col, col)\n",
    "    plot += '<text x=\"%s\" y=\"%s\" font-family=\"Verdana\" font-size=\"17\">%s</text>\\n' %(x_coord + 15, y + 12, \"Intergenic\")\n",
    "\n",
    "    # Adding scale bars:\n",
    "    length = 63\n",
    "    x = 40\n",
    "    y = 160\n",
    "    tick = 5\n",
    "\n",
    "    plot += '<text x=\"%s\" y=\"%s\" font-family=\"Verdana\" text-anchor=\"middle\"  font-size=\"15\">15kbp</text>\\n' % (length/2 + x, y - 10)\n",
    "    plot += '<line x1=\"%s\" y1=\"%s\" x2=\"%s\" y2=\"%s\" style=\"stroke:black;stroke-width:3\" />\\n' %(x,y,x + length,y)\n",
    "    plot += '<line x1=\"%s\" y1=\"%s\" x2=\"%s\" y2=\"%s\" style=\"stroke:black;stroke-width:3\" />\\n' %(x,y-tick,x,y+tick)\n",
    "    plot += '<line x1=\"%s\" y1=\"%s\" x2=\"%s\" y2=\"%s\" style=\"stroke:black;stroke-width:3\" />\\n' %(x + length,y-tick,x + length,y+tick)\n",
    "\n",
    "    # Adding verical units: 42 units for 500kbp\n",
    "    # Get the length of one row: \n",
    "    length_row = len(df[df.Row_no == 3]) * 480\n",
    "    length_bar = 50\n",
    "    bar_size_genome = round(length_bar / 2 * length_row/1e6, 2)\n",
    "\n",
    "    x = 140\n",
    "    y = 160\n",
    "    plot += '<text x=\"%s\" y=\"%s\" font-family=\"Verdana\" font-size=\"15\">%sMbp</text>\\n' % (\n",
    "        x + 15, y, bar_size_genome)\n",
    "    plot += '<line x1=\"%s\" y1=\"%s\" x2=\"%s\" y2=\"%s\" style=\"stroke:black;stroke-width:3\" />\\n' % (\n",
    "        x, y-length_bar/2, x, y + length_bar/2)\n",
    "    plot += '<line x1=\"%s\" y1=\"%s\" x2=\"%s\" y2=\"%s\" style=\"stroke:black;stroke-width:3\" />\\n' % (\n",
    "        x-tick,y-length_bar/2,x+tick,y-length_bar/2)\n",
    "    plot += '<line x1=\"%s\" y1=\"%s\" x2=\"%s\" y2=\"%s\" style=\"stroke:black;stroke-width:3\" />\\n' % (\n",
    "        x-tick,y + length_bar/2,x+tick,y + length_bar/2)\n",
    "\n",
    "    # Adding a circle to the bottom of the legend:\n",
    "    circle_r = 5\n",
    "    circle_x = 40 \n",
    "    circle_y = 210\n",
    "    plot += '<circle cx=\"%s\" cy=\"%s\" r=\"%s\" stroke=\"black\" stroke-width=\"1\" fill=\"black\" />\\n' % (\n",
    "            circle_x, circle_y, circle_r)\n",
    "    plot += '<text x=\"%s\" y=\"%s\" font-family=\"Verdana\" font-size=\"15\">%s</text>\\n' % (\n",
    "            circle_x + 20, circle_y + 5, \"Known GWAS signals\")\n",
    "\n",
    "    # Closing legend\n",
    "    plot += '</g>'\n",
    "\n",
    "    #f.write('</svg>\\n</body>\\n</html>\\n')\n",
    "    plot += '</svg>\\n'\n",
    "    #f.close()\n",
    "\n",
    "    fout = open('%s_chr%s.png' % (filename, chromosome),'w')\n",
    "    cairosvg.svg2png(bytestring=plot,write_to=fout)\n",
    "    fout.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing chromosome 1\n",
      "Processing chromosome 2\n",
      "Processing chromosome 3\n",
      "Processing chromosome 4\n",
      "Processing chromosome 5\n",
      "Processing chromosome 6\n",
      "Processing chromosome 7\n",
      "Processing chromosome 8\n",
      "Processing chromosome 9\n",
      "Processing chromosome 10\n",
      "Processing chromosome 11\n",
      "Processing chromosome 12\n",
      "Processing chromosome 13\n",
      "Processing chromosome 14\n",
      "Processing chromosome 15\n",
      "Processing chromosome 16\n",
      "Processing chromosome 17\n",
      "Processing chromosome 18\n",
      "Processing chromosome 19\n",
      "Processing chromosome 20\n",
      "Processing chromosome 21\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Now we loop through all autosomes and generate a plot.\n",
    "In the meanwhile, the dataframes are also saved, and upon a future step,\n",
    "they can be read back withouth calculating them from the scratch.\n",
    "'''\n",
    "for chromosome in range(1, 22):\n",
    "    print \"Processing chromosome %s\" % (chromosome)\n",
    "\n",
    "    # Process chromosome data, assign colors, combine into dataframe:\n",
    "    df = ProcessChromosome(chromosome)\n",
    "\n",
    "    # Pickling dataframes:\n",
    "    filename = outputDir % chromosome\n",
    "    output = open(filename, 'wb')\n",
    "    pickle.dump(df, output)\n",
    "    \n",
    "    # If we jus want ot read df from file:\n",
    "\n",
    "    # Add colors to datafrome:\n",
    "    df = color_chromosome(df, chromosome)\n",
    "    plot_chromosome(df, chromosome, \"plot\", 600)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ds26/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (0,1,2,3) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'itemRgb'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-222-3373c85c303e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;31m# Assign features to each chunk:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 17\u001b[0;31m \u001b[0mGENCODE_chunks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mGC_INT\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'score'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m'exon'\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0;34m'exon'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitemRgb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;34m'gene'\u001b[0m \u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     18\u001b[0m \u001b[0mGENCODE_chunks\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"GENCODE\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0mchr_dataf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'GENCODE'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'intergenic'\u001b[0m \u001b[0;31m# By default, all GENCODE values are intergenic\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ds26/anaconda/lib/python3.6/site-packages/pandas/core/groupby.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, func, *args, **kwargs)\u001b[0m\n\u001b[1;32m    692\u001b[0m         \u001b[0;31m# ignore SettingWithCopy here in case the user mutates\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    693\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0moption_context\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'mode.chained_assignment'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 694\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_python_apply_general\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    695\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    696\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_python_apply_general\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ds26/anaconda/lib/python3.6/site-packages/pandas/core/groupby.py\u001b[0m in \u001b[0;36m_python_apply_general\u001b[0;34m(self, f)\u001b[0m\n\u001b[1;32m    696\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_python_apply_general\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    697\u001b[0m         keys, values, mutated = self.grouper.apply(f, self._selected_obj,\n\u001b[0;32m--> 698\u001b[0;31m                                                    self.axis)\n\u001b[0m\u001b[1;32m    699\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    700\u001b[0m         return self._wrap_applied_output(\n",
      "\u001b[0;32m/Users/ds26/anaconda/lib/python3.6/site-packages/pandas/core/groupby.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, f, data, axis)\u001b[0m\n\u001b[1;32m   1609\u001b[0m             \u001b[0;31m# group might be modified\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1610\u001b[0m             \u001b[0mgroup_axes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_get_axes\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgroup\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1611\u001b[0;31m             \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgroup\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1612\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0m_is_indexed_like\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mres\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgroup_axes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1613\u001b[0m                 \u001b[0mmutated\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-222-3373c85c303e>\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;31m# Assign features to each chunk:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 17\u001b[0;31m \u001b[0mGENCODE_chunks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mGC_INT\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'score'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m'exon'\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0;34m'exon'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitemRgb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;34m'gene'\u001b[0m \u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     18\u001b[0m \u001b[0mGENCODE_chunks\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"GENCODE\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0mchr_dataf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'GENCODE'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'intergenic'\u001b[0m \u001b[0;31m# By default, all GENCODE values are intergenic\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ds26/anaconda/lib/python3.6/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   2742\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mname\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2743\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2744\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2745\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2746\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'itemRgb'"
     ]
    }
   ],
   "source": [
    "# Reading datafile:\n",
    "chr_name = 2\n",
    "dataFile = Chromosome_file_loc % chr_name\n",
    "chr_dataf = pd.read_csv(dataFile, compression='gzip', header=None, sep='\\t', quotechar='\"')\n",
    "\n",
    "# Reading GENCODE file:\n",
    "GENCODE_bed = pybedtools.BedTool(GENCODE_file_loc)\n",
    "\n",
    "# Run intersectbed:\n",
    "chr_data_bed = pybedtools.BedTool(dataFile)\n",
    "\n",
    "# Get intersecting GENCODE features:\n",
    "GencodeIntersect = chr_data_bed.intersect(GENCODE_bed, wa = True, wb = True)\n",
    "GC_INT = pybedtools.bedtool.BedTool.to_dataframe(GencodeIntersect)\n",
    "\n",
    "# Assign features to each chunk:\n",
    "GENCODE_chunks = GC_INT.groupby('score').apply(lambda x: 'exon' if 'exon' in x.itemRgb.unique() else 'gene' )\n",
    "GENCODE_chunks.name = \"GENCODE\"\n",
    "chr_dataf['GENCODE'] = 'intergenic' # By default, all GENCODE values are intergenic\n",
    "chr_dataf.GENCODE.update(GENCODE_chunks) # This value will be overwritten if overlaps with exon or gene\n",
    "\n",
    "# Assinging column number:\n",
    "row_count = 800\n",
    "col_count = int(chr_dataf.shape[0] / rows)\n",
    "chr_dataf['col_frac'] = chr_dataf[4] % col_count / float(col_count)\n",
    "chr_dataf.head()\n",
    "\n",
    "# Get colors based on GENCODE feature:\n",
    "colors_GENCODE = {\n",
    "    'intergenic': linear_gradient('#42b79a', finish_hex='#42b79a', n=20), # gray\n",
    "    'exon': linear_gradient('#CDCD00', n=20), # Purple \n",
    "    'gene': linear_gradient('#4278b7', n=20)} # Goldenrod\n",
    "chr_dataf['color'] = chr_dataf.apply(lambda x: colors_GENCODE[x['GENCODE']][int(x[3]*20)], axis = 1)\n",
    "\n",
    "# Get the colors darker based on column number:\n",
    "chr_dataf['color'] = chr_dataf.apply(color_darkener, axis = 1)\n",
    "\n",
    "#plot_chromosome(chr_dataf, rows=row_count)\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   0      1      2         3  4     GENCODE  col_frac    color\n",
      "0  2  10020  10500  0.612500  1  intergenic  0.001597  #b9e4d9\n",
      "1  2  10500  10980  0.662500  2  intergenic  0.003195  #c3e8df\n",
      "2  2  10980  11460  0.787500  3  intergenic  0.004792  #d7efe9\n",
      "3  2  11460  11940  0.645833  4  intergenic  0.006390  #b9e4d9\n",
      "4  2  11940  12420  0.339583  5  intergenic  0.007987  #7dcdb9\n"
     ]
    }
   ],
   "source": [
    "def plot_chromosome(df,rows = 600):\n",
    "    '''\n",
    "    Genomic data processed, colors have been assigned to each chunk.\n",
    "    Now create an svg image, then render it in a png.\n",
    "    '''\n",
    "    width = int(chr_dataf.shape[0] / rows)\n",
    "\n",
    "    #chromosome_name = chromosome\n",
    "    # Although I decided to focus on the autosomal chromosomes, the script is ready to deal with\n",
    "    # the sex chromosomes.\n",
    "    #if chromosome_name == 23: chromosome_name = \"X\"\n",
    "    #if chromosome_name == 24: chromosome_name = \"Y\"        \n",
    "\n",
    "    # The size of box which represent a single chunk of the genome:\n",
    "    box_width = 2\n",
    "    box_height = 2\n",
    "\n",
    "    # svg initiated:\n",
    "    plot = '<svg width=\"%s\" height=\"%s\" viewBox=\"0 0 %s %s\">\\n' \\\n",
    "        % (width*(box_width+1), rows*(box_height+1), width*(box_width+1), rows*(box_height+1))\n",
    "\n",
    "    # Plotting genome based on GC content and gene annotation:\n",
    "    for index, color in enumerate(df.color):\n",
    "        # Get x,y coordinates based on index:\n",
    "        y = int(index / width) * (box_height + 1)\n",
    "        x = (index - int(index / width)*width)*(box_width+1)\n",
    "\n",
    "        # Generate svg line:\n",
    "        plot +='<rect x=\"%s\" y=\"%s\" width=\"%s\" height=\"%s\" style=\"stroke-width:1;stroke:%s; fill: %s\" />\\n' % (x, y, box_width + 1 , box_height + 1, color, color)\n",
    "        \n",
    "    # looking through gwas annotation:\n",
    "    #for index, gwas in enumerate(df.GWAS):\n",
    "    #    if gwas == 1:\n",
    "    #        # based on the index of the given line, we get the x,y coordinates:\n",
    "    #        y = (index / width) * (box_height + 1) + 1.5\n",
    "    #        x = (index - (index / width)*width)*(box_width+1) + 1.5\n",
    "\n",
    "    #        # Drawing circle around the defined point:\n",
    "    #        plot += '<circle cx=\"%s\" cy=\"%s\" r=\"3\" stroke=\"%s\" stroke-width=\"1\" fill=\"%s\" />\\n' % (x, y, \"black\", \"black\")\n",
    "\n",
    "    plot += '</svg>\\n'# Terminate svg \n",
    "    \n",
    "    #f.close()\n",
    "\n",
    "    fout = open('cicaful.svg','w')\\\n",
    "    # Saving svg into file:\n",
    "    fout.write(plot)\n",
    "    #cairosvg.svg2png(bytestring=plot,write_to='cicaful.png')\n",
    "    fout.close()\n",
    "    \n",
    "print(chr_dataf.head())\n",
    "#plot_chromosome(df, chromosome, filename, rows = 600):\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![test](cicaful.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "220    #e8ecec\n",
       "221    #e1e7e7\n",
       "222    #e4e9e9\n",
       "223    #e4e9e9\n",
       "224    #e4e9e9\n",
       "225    #dd99ff\n",
       "226    #ffd775\n",
       "227    #d685ff\n",
       "228    #dd99ff\n",
       "229    #ffdb80\n",
       "dtype: object"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def hex_to_RGB(hex):\n",
    "    ''' \"#FFFFFF\" -> [255,255,255] '''\n",
    "    # Pass 16 to the integer function for change of base\n",
    "    return [int(hex[i:i+2], 16) for i in range(1,6,2)]\n",
    "\n",
    "\n",
    "def RGB_to_hex(RGB):\n",
    "    ''' [255,255,255] -> \"#FFFFFF\" '''\n",
    "    # Components need to be integers for hex to make sense\n",
    "    RGB = [int(x) for x in RGB]\n",
    "    return \"#\"+\"\".join([\"0{0:x}\".format(v) if v < 16 else\n",
    "            \"{0:x}\".format(v) for v in RGB])\n",
    "\n",
    "def linear_gradient(start_hex, finish_hex=\"#FFFFFF\", n=10):\n",
    "    ''' \n",
    "    returns a gradient list of (n) colors between\n",
    "    two hex colors. start_hex and finish_hex\n",
    "    should be the full six-digit color string,\n",
    "    inlcuding the sharp sign (eg \"#FFFFFF\") \n",
    "    '''\n",
    "    # Starting and ending colors in RGB form\n",
    "    s = hex_to_RGB(start_hex)\n",
    "    f = hex_to_RGB(finish_hex)\n",
    "    \n",
    "    # Initilize a list of the output colors with the starting color\n",
    "    RGB_list = [start_hex]\n",
    "    \n",
    "    # Calcuate a color at each evenly spaced value of t from 1 to n\n",
    "    for t in range(1, n):\n",
    "    \n",
    "        # Interpolate RGB vector for color at the current value of t\n",
    "        curr_vector = [\n",
    "            int(s[j] + (float(t)/(n-1))*(f[j]-s[j]))\n",
    "            for j in range(3)\n",
    "        ]\n",
    "\n",
    "        # Add it to our list of output colors\n",
    "        RGB_list.append(RGB_to_hex(curr_vector))\n",
    "\n",
    "    return RGB_list\n",
    "\n",
    "def color_darkener(row):\n",
    "    '''\n",
    "    Once the colors are assigned, we make the colors darker for those columns \n",
    "    that are at the end of the plot.\n",
    "    '''\n",
    "    \n",
    "    color = row['color']\n",
    "    col_frac = row['col_frac']\n",
    "    \n",
    "    # Built in parameters:\n",
    "    threshold = 0.7 # at which column the darkening starts\n",
    "    max_diff_value = 0.15 # The max value of darkening\n",
    "    \n",
    "    # Ok, we have the color, now based on the column we make it a bit darker:\n",
    "    if col_frac > threshold:\n",
    "        diff = (col_frac - threshold)/(1 - threshold)\n",
    "        factor = 1 - max_diff_value*diff \n",
    "\n",
    "        # Get rgb code of the hexa code:\n",
    "        rgb_code = hex_to_RGB(color)\n",
    "\n",
    "        # Get the hls code of the rgb:\n",
    "        hls_code = colorsys.rgb_to_hls(rgb_code[0]/float(255), \n",
    "                                       rgb_code[1]/float(255), \n",
    "                                       rgb_code[2]/float(255))\n",
    "\n",
    "        # Get the modifed rgb code:\n",
    "        new_rgb = colorsys.hls_to_rgb(hls_code[0], \n",
    "                                      hls_code[1]*factor, \n",
    "                                      hls_code[2])\n",
    "\n",
    "        # Get the modifed hexacode:\n",
    "        color = RGB_to_hex([new_rgb[0] * 255,\n",
    "                           new_rgb[1] * 255,\n",
    "                           new_rgb[2] * 255])\n",
    "    return(color)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "int() argument must be a string, a bytes-like object or a number, not 'list'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-225-eb9694494c37>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mGWAS_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mGWAS_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchrom\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchr_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: int() argument must be a string, a bytes-like object or a number, not 'list'"
     ]
    }
   ],
   "source": [
    "GWAS_df[GWAS_df.chrom == str(chr_name)].start.tolist()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fin = open('cicaful.test.svg','r')\n",
    "plot = fin.read()\n",
    "fin.close()\n",
    "\n",
    "# This value comes from the sh script:\n",
    "window = 480\n",
    "pixel = 3\n",
    "row_count = 800\n",
    "\n",
    "#GWAS_file_loc\n",
    "GWAS_file = pybedtools.BedTool(GWAS_file_loc)\n",
    "GWAS_df = pybedtools.bedtool.BedTool.to_dataframe(GWAS_file)\n",
    "GWAS_df[GWAS_df.chrom == str(chr_name)].start\n",
    "\n",
    "# Looping through all the GWAS hits:\n",
    "for pos in GWAS_df[GWAS_df.chrom == str(chr_name)].start.unique().tolist():\n",
    "    # Based on the position, calculate chunk:\n",
    "    chunk = int(int(pos)/window)\n",
    "    \n",
    "    # Based on the chunk no and the pixel size and row count, let's get the coordinates:\n",
    "    y = int(chunk / row_count) * (pixel)\n",
    "    x = (chunk - int(chunk / row_count)*row_count)*(pixel)\n",
    "    \n",
    "    # Draw point:\n",
    "    plot += '<circle cx=\"%s\" cy=\"%s\" r=\"3\" stroke=\"%s\" stroke-width=\"1\" fill=\"%s\" />\\n' % \\\n",
    "        (x, y, \"black\", \"black\")\n",
    "        \n",
    "plot += '</svg>\\n'# Terminate svg \n",
    "\n",
    "fout = open('cicaful_GWAS.svg','w')\\\n",
    "# Saving svg into file:\n",
    "fout.write(plot)\n",
    "#cairosvg.svg2png(bytestring=plot,write_to='cicaful_GWAS.png')\n",
    "fout.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk: 26228\n",
      "X: 1884, Y: 96\n"
     ]
    }
   ],
   "source": [
    "pos = 11802720\n",
    "window = 400\n",
    "chunk = int(int(pos)/window)\n",
    "print(\"Chunk: %s\" % chunk)\n",
    "y = int(chunk / row_count) * (pixel)\n",
    "x = (chunk - int(chunk / row_count)*row_count)*(pixel)\n",
    "print(\"X: %s, Y: %s\" % (x,y))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>GC_ratio</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>GENCODE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.0</td>\n",
       "      <td>10000.0</td>\n",
       "      <td>10500.0</td>\n",
       "      <td>0.612</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>10500.0</td>\n",
       "      <td>11000.0</td>\n",
       "      <td>0.668</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.0</td>\n",
       "      <td>11000.0</td>\n",
       "      <td>11500.0</td>\n",
       "      <td>0.770</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.0</td>\n",
       "      <td>11500.0</td>\n",
       "      <td>12000.0</td>\n",
       "      <td>0.632</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2.0</td>\n",
       "      <td>12000.0</td>\n",
       "      <td>12500.0</td>\n",
       "      <td>0.348</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr    start      end  GC_ratio  x  y     GENCODE\n",
       "0  2.0  10000.0  10500.0     0.612  0  0  intergenic\n",
       "1  2.0  10500.0  11000.0     0.668  1  0  intergenic\n",
       "2  2.0  11000.0  11500.0     0.770  2  0  intergenic\n",
       "3  2.0  11500.0  12000.0     0.632  3  0  intergenic\n",
       "4  2.0  12000.0  12500.0     0.348  4  0  intergenic"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def generate_xy(pos, min_pos, chunk_size, width):\n",
    "    '''\n",
    "    Generating x,y coordinates from genomic position and \n",
    "    the provided plot width or chunk size.\n",
    "\n",
    "    pos - genomic position\n",
    "    pixel - the size of the point taken up by one chunk.\n",
    "    width - number of chunks in one row\n",
    "    chunk_size - the number of basepairs pooled together in one chunk.\n",
    "    '''\n",
    "    pos = pos - min_pos\n",
    "    chunk = int(int(pos)/chunk_size)\n",
    "    print(\"Chunk: %s\" % chunk)\n",
    "    y = int(chunk / row_count)\n",
    "    x = (chunk - int(chunk / row_count)*row_count)\n",
    "    print(\"X: %s, Y: %s\" % (x,y))\n",
    "\n",
    "chr_dataf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Cleaning up the code\n",
    "\n",
    "\n",
    "* input parameters:\n",
    "    * chromosome name\n",
    "    * chunk size\n",
    "    * axis\n",
    "    * dimension of the plot\n",
    "* Open input files\n",
    "* Assign x,y columns to each chunk\n",
    "* Assign GENCODE feature to each chunk.\n",
    "* Assign color to each chunk\n",
    "* Darken color if neccessary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>thickStart</th>\n",
       "      <th>thickEnd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>38500</td>\n",
       "      <td>39000</td>\n",
       "      <td>0.340</td>\n",
       "      <td>2</td>\n",
       "      <td>38814</td>\n",
       "      <td>41627</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>38500</td>\n",
       "      <td>39000</td>\n",
       "      <td>0.340</td>\n",
       "      <td>2</td>\n",
       "      <td>38814</td>\n",
       "      <td>46870</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>39000</td>\n",
       "      <td>39500</td>\n",
       "      <td>0.416</td>\n",
       "      <td>2</td>\n",
       "      <td>38814</td>\n",
       "      <td>41627</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>39000</td>\n",
       "      <td>39500</td>\n",
       "      <td>0.416</td>\n",
       "      <td>2</td>\n",
       "      <td>38814</td>\n",
       "      <td>46870</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>39500</td>\n",
       "      <td>40000</td>\n",
       "      <td>0.378</td>\n",
       "      <td>2</td>\n",
       "      <td>38814</td>\n",
       "      <td>41627</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom  start    end   name  score  strand  thickStart thickEnd\n",
       "0      2  38500  39000  0.340      2   38814       41627     exon\n",
       "1      2  38500  39000  0.340      2   38814       46870     gene\n",
       "2      2  39000  39500  0.416      2   38814       41627     exon\n",
       "3      2  39000  39500  0.416      2   38814       46870     gene\n",
       "4      2  39500  40000  0.378      2   38814       41627     exon"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Importig libraries:\n",
    "'''\n",
    "\n",
    "import gzip # For reading data\n",
    "import pybedtools # For finding overlap between our chunks and genes and GWAS signals\n",
    "import pandas # data handling\n",
    "import numpy as np # Working with large arrays\n",
    "import colorsys # Generate color gradient.\n",
    "import cairosvg # Converting svg to image\n",
    "import pickle # Saving dataframes to disk\n",
    "import os.path # checking if the datafies are already there.\n",
    "\n",
    "'''\n",
    "Input files and folder + checking if they are present\n",
    "'''\n",
    "\n",
    "workingDir = os.getcwd()\n",
    "Chromosome_file_loc = workingDir + \"/data/Processed_chr%s.bed.gz\"\n",
    "GWAS_file_loc = workingDir + \"/data/processed_GWAS.bed.gz\"\n",
    "GENCODE_file_loc = workingDir + \"/data/GENCODE.merged.bed.gz\"\n",
    "outputDir = workingDir + \"/processed_data/dataframe_chr%s.pkl\"\n",
    "\n",
    "if not os.path.isfile(GWAS_file_loc):\n",
    "    print (\"GWAS file is missing. Run `prepare_data.sh` first!\")\n",
    "if not os.path.isfile(GENCODE_file_loc):\n",
    "    print (\"GENCODE file is missing. Run `prepare_data.sh` first!\")\n",
    "if not os.path.isfile(Chromosome_file_loc % 11):\n",
    "    print (\"Processed chromosome files are missing! Run `prepare_data.sh` first!\")\n",
    "\n",
    "# These values are read from the command line once the script is wrapped:\n",
    "axis = 1\n",
    "dimension = 400 \n",
    "chromosome = 2 \n",
    "\n",
    "# Reading datafile:\n",
    "dataFile = Chromosome_file_loc % chromosome\n",
    "chr_dataf = pd.read_csv(dataFile, compression='gzip', sep='\\t', quotechar='\"')\n",
    "\n",
    "# Based on the difference between the stat and end positions, we get the chunk size:\n",
    "chunk_size = chr_dataf.head(1).apply(lambda x: x['end'] - x['start'], axis=1)\n",
    "min_pos  = chr_dataf.start.min()\n",
    "\n",
    "# Get the width (number of chunks plotted in one row)\n",
    "width = dimension # We are dealing with fixed column numbers\n",
    "if axis == 0:  width = int(chr_dataf.shape[0]/dimension) # We are dealing with fixed number of rows.\n",
    "\n",
    "# Adding plot coordinates to the dataframe:\n",
    "chr_dataf = chr_dataf.apply(generate_xy, args = (min_pos, chunk_size, width), axis =1)\n",
    "\n",
    "# Reading GENCODE file:\n",
    "GENCODE_bed = pybedtools.BedTool(GENCODE_file_loc)\n",
    "\n",
    "# Run intersectbed:\n",
    "chr_data_bed = pybedtools.BedTool(dataFile)\n",
    "\n",
    "# Get intersecting GENCODE features:\n",
    "GencodeIntersect = chr_data_bed.intersect(GENCODE_bed, wa = True, wb = True)\n",
    "GC_INT = pybedtools.bedtool.BedTool.to_dataframe(GencodeIntersect)\n",
    "\n",
    "# Assign features to each chunk:\n",
    "#GENCODE_chunks = GC_INT.groupby('score').apply(lambda x: 'exon' if 'exon' in x.itemRgb.unique() else 'gene' )\n",
    "#GENCODE_chunks.name = \"GENCODE\"\n",
    "#chr_dataf['GENCODE'] = 'intergenic' # By default, all GENCODE values are intergenic\n",
    "#chr_dataf.GENCODE.update(GENCODE_chunks) # This value will be overwritten if overlaps with exon or gene\n",
    "#chr_dataf.head()\n",
    "\n",
    "# Assinging column number:\n",
    "#row_count = 800\n",
    "#col_count = int(chr_dataf.shape[0] / rows)\n",
    "#chr_dataf['col_frac'] = chr_dataf[4] % col_count / float(col_count)\n",
    "#chr_dataf.head()\n",
    "GC_INT.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "GENCODE_chunks = GC_INT.groupby('start').apply(lambda x: 'exon' if 'exon' in x.thickEnd.unique() else 'gene' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['intergenic', 'exon', 'gene'], dtype=object)"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GENCODE_chunks.name = \"GENCODE\"\n",
    "chr_dataf['GENCODE'] = 'intergenic'\n",
    "chr_dataf.GENCODE.update(GENCODE_chunks)\n",
    "chr_dataf.GENCODE.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>GC_ratio</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>GENCODE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>10500.0</td>\n",
       "      <td>11000.0</td>\n",
       "      <td>0.668</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.0</td>\n",
       "      <td>11000.0</td>\n",
       "      <td>11500.0</td>\n",
       "      <td>0.770</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.0</td>\n",
       "      <td>11500.0</td>\n",
       "      <td>12000.0</td>\n",
       "      <td>0.632</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2.0</td>\n",
       "      <td>12000.0</td>\n",
       "      <td>12500.0</td>\n",
       "      <td>0.348</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2.0</td>\n",
       "      <td>12500.0</td>\n",
       "      <td>13000.0</td>\n",
       "      <td>0.422</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr    start      end  GC_ratio  x  y     GENCODE\n",
       "1  2.0  10500.0  11000.0     0.668  1  0  intergenic\n",
       "2  2.0  11000.0  11500.0     0.770  2  0  intergenic\n",
       "3  2.0  11500.0  12000.0     0.632  3  0  intergenic\n",
       "4  2.0  12000.0  12500.0     0.348  4  0  intergenic\n",
       "5  2.0  12500.0  13000.0     0.422  5  0  intergenic"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def generate_xy(df, min_pos, chunk_size, width):\n",
    "    '''\n",
    "    Generating x,y coordinates from genomic position and\n",
    "    the provided plot width or chunk size.\n",
    "\n",
    "    pos - genomic position\n",
    "    pixel - the size of the point taken up by one chunk.\n",
    "    width - number of chunks in one row\n",
    "    chunk_size - the number of basepairs pooled together in one chunk.\n",
    "    '''\n",
    "    pos = df['start'] - min_pos\n",
    "    chunk = int(int(pos)/chunk_size)\n",
    "    df[\"x\"] = (chunk - int(chunk / width)*width)\n",
    "    df[\"y\"] = int(chunk / width)\n",
    "    return (df)\n",
    "\n",
    "test_df = chr_dataf.ix[1:10000]\n",
    "test_df = test_df.apply(generate_xy, args = (min_pos, chunk_size, width), axis =1)\n",
    "test_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   chr    start      end  GC_ratio  x  y     GENCODE\n",
      "0  2.0  10000.0  10500.0     0.612  0  0  intergenic\n",
      "1  2.0  10500.0  11000.0     0.668  1  0  intergenic\n",
      "2  2.0  11000.0  11500.0     0.770  2  0  intergenic\n",
      "3  2.0  11500.0  12000.0     0.632  3  0  intergenic\n",
      "4  2.0  12000.0  12500.0     0.348  4  0  intergenic\n"
     ]
    }
   ],
   "source": [
    "#print(\"[Info %s] Assigning colors to each chunk.\" % (get_now()))\n",
    "print(chr_dataf.head())\n",
    "colors_GENCODE = {\n",
    "    'intergenic': linear_gradient('#42b79a', n=20), # gray\n",
    "    'exon': linear_gradient('#CDCD00', n=20), # Purple\n",
    "    'gene': linear_gradient('#4278b7', n=20)} # Goldenrod"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>GC_ratio</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>GENCODE</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>10500.0</td>\n",
       "      <td>11000.0</td>\n",
       "      <td>0.668</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>#c3e8df</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.0</td>\n",
       "      <td>11000.0</td>\n",
       "      <td>11500.0</td>\n",
       "      <td>0.770</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>#d7efe9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.0</td>\n",
       "      <td>11500.0</td>\n",
       "      <td>12000.0</td>\n",
       "      <td>0.632</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>#b9e4d9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2.0</td>\n",
       "      <td>12000.0</td>\n",
       "      <td>12500.0</td>\n",
       "      <td>0.348</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>#7dcdb9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2.0</td>\n",
       "      <td>12500.0</td>\n",
       "      <td>13000.0</td>\n",
       "      <td>0.422</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>#91d5c4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr    start      end  GC_ratio  x  y     GENCODE    color\n",
       "1  2.0  10500.0  11000.0     0.668  1  0  intergenic  #c3e8df\n",
       "2  2.0  11000.0  11500.0     0.770  2  0  intergenic  #d7efe9\n",
       "3  2.0  11500.0  12000.0     0.632  3  0  intergenic  #b9e4d9\n",
       "4  2.0  12000.0  12500.0     0.348  4  0  intergenic  #7dcdb9\n",
       "5  2.0  12500.0  13000.0     0.422  5  0  intergenic  #91d5c4"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df['color'] = test_df.apply(lambda x: colors_GENCODE[x['GENCODE']][int(x[3]*20)], axis = 1)\n",
    "test_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def color_darkener(row, width, threshold, max_diff_value):\n",
    "    '''\n",
    "    width - how many chunks do we have in one line.\n",
    "    threshold - at which column the darkening starts\n",
    "    max_diff_value - the max value of darkening\n",
    "    \n",
    "    Once the colors are assigned, we make the colors darker for those columns\n",
    "    that are at the end of the plot.\n",
    "    '''\n",
    "\n",
    "    color = row['color']\n",
    "    col_frac = row['x'] / width\n",
    "\n",
    "    # Ok, we have the color, now based on the column we make it a bit darker:\n",
    "    if col_frac > threshold:\n",
    "        diff = (col_frac - threshold)/(1 - threshold)\n",
    "        factor = 1 - max_diff_value*diff\n",
    "\n",
    "        # Get rgb code of the hexa code:\n",
    "        rgb_code = hex_to_RGB(color)\n",
    "\n",
    "        # Get the hls code of the rgb:\n",
    "        hls_code = colorsys.rgb_to_hls(rgb_code[0]/float(255),\n",
    "                                       rgb_code[1]/float(255),\n",
    "                                       rgb_code[2]/float(255))\n",
    "\n",
    "        # Get the modifed rgb code:\n",
    "        new_rgb = colorsys.hls_to_rgb(hls_code[0],\n",
    "                                      hls_code[1]*factor,\n",
    "                                      hls_code[2])\n",
    "\n",
    "        # Get the modifed hexacode:\n",
    "        color = RGB_to_hex([new_rgb[0] * 255,\n",
    "                           new_rgb[1] * 255,\n",
    "                           new_rgb[2] * 255])\n",
    "    return(color)\n",
    "\n",
    "start = 0.75\n",
    "threshold = 0.15\n",
    "test_df['color'] = test_df.apply(color_darkener, axis = 1, args=(width, start, threshold))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def generate_xy(df, min_pos, chunk_size, width):\n",
    "    '''\n",
    "    Generating x,y coordinates from genomic position and\n",
    "    the provided plot width or chunk size.\n",
    "\n",
    "    pos - genomic position\n",
    "    pixel - the size of the point taken up by one chunk.\n",
    "    width - number of chunks in one row\n",
    "    chunk_size - the number of basepairs pooled together in one chunk.\n",
    "    '''\n",
    "    pos = int(df['start']) - min_pos\n",
    "    chunk = int(int(pos)/chunk_size)\n",
    "    df[\"x\"] = (chunk - int(chunk / width)*width)\n",
    "    df[\"y\"] = int(chunk / width)\n",
    "    return (df)\n",
    "\n",
    "GWAS_file = pybedtools.BedTool(GWAS_file_loc)\n",
    "full_GWAS = pybedtools.bedtool.BedTool.to_dataframe(GWAS_file)\n",
    "GWAS_df = full_GWAS[full_GWAS.chrom == str(chromosome)]\n",
    "\n",
    "chr_dataf.start.min()\n",
    "chunk_size = 500\n",
    "width = 400\n",
    "GWAS_df = GWAS_df.apply(generate_xy, args = (min_pos, chunk_size, width), axis =1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Info 01:08:04] reading file /Users/ds26/Projects/GenomePlotter/data/Processed_chr22.bed.gz... \n",
      "[Info 01:08:04] Number of chunks read: 112929\n",
      "[Info 01:08:04] Chunk size: 450.0 bp, plot width: 300\n",
      "[Info 01:08:04] Calculating plot coordinates for each chunk.\n",
      "        chr   start     end  GC_ratio  x  y\n",
      "start                                      \n",
      "0      22.0     0.0   450.0       NaN  0  0\n",
      "450    22.0   450.0   900.0       NaN  1  0\n",
      "900    22.0   900.0  1350.0       NaN  2  0\n",
      "1350   22.0  1350.0  1800.0       NaN  3  0\n",
      "1800   22.0  1800.0  2250.0       NaN  4  0\n"
     ]
    }
   ],
   "source": [
    "from time import gmtime, strftime\n",
    "import sys \n",
    "\n",
    "def get_now():\n",
    "    return strftime(\"%H:%M:%S\", gmtime())\n",
    "\n",
    "\n",
    "workingDir = os.getcwd()\n",
    "Chromosome_file_loc = workingDir + \"/data/Processed_chr%s.bed.gz\"\n",
    "GWAS_file_loc = workingDir + \"/data/processed_GWAS.bed.gz\"\n",
    "GENCODE_file_loc = workingDir + \"/data/GENCODE.merged.bed.gz\"\n",
    "outputDir = workingDir + \"/processed_data/dataframe_chr%s.pkl\"\n",
    "\n",
    "if not os.path.isfile(GWAS_file_loc):\n",
    "    print (\"GWAS file is missing. Run `prepare_data.sh` first!\")\n",
    "if not os.path.isfile(GENCODE_file_loc):\n",
    "    print (\"GENCODE file is missing. Run `prepare_data.sh` first!\")\n",
    "if not os.path.isfile(Chromosome_file_loc % 11):\n",
    "    print (\"Processed chromosome files are missing! Run `prepare_data.sh` first!\")\n",
    "\n",
    "# These values are read from the command line once the script is wrapped:\n",
    "axis = 1\n",
    "dimension = 300 \n",
    "chromosome = 22\n",
    "\n",
    "# Reading datafile:\n",
    "dataFile = Chromosome_file_loc % chromosome\n",
    "print(\"[Info %s] reading file %s... \" % (get_now(), dataFile))\n",
    "chr_dataf = pd.read_csv(dataFile, compression='gzip', sep='\\t', quotechar='\"')\n",
    "\n",
    "# If we want to do test, we can restrict the dataframe to n lines:\n",
    "if len(sys.argv) == 6:\n",
    "    chr_dataf = chr_dataf.ix[0:int(sys.argv[5])]\n",
    "    print(\"[Info %s] Test mode is on. The first %s rows will be kept.\" %(get_now(), sys.argv[5]))\n",
    "\n",
    "# the start position will be the new index:\n",
    "chr_dataf = chr_dataf.set_index('start', drop = False)\n",
    "\n",
    "chunk_size = chr_dataf.head(1).apply(lambda x: x['end'] - x['start'], axis = 1).tolist()[0]\n",
    "min_pos  = chr_dataf.start.min()\n",
    "print(\"[Info %s] Number of chunks read: %s\" %(get_now(), chr_dataf.shape[0]))\n",
    "\n",
    "# Get the width (number of chunks plotted in one row)\n",
    "width = dimension # We are dealing with fixed column numbers\n",
    "if axis == 2:  width = int(chr_dataf.shape[0]/dimension) # We are dealing with fixed number of rows.\n",
    "\n",
    "# Adding plot coordinates to the dataframe:\n",
    "print(\"[Info %s] Chunk size: %s bp, plot width: %s\" %(get_now(), chunk_size, width))\n",
    "print(\"[Info %s] Calculating plot coordinates for each chunk.\" %(get_now()))\n",
    "chr_dataf = chr_dataf.apply(generate_xy, args = (min_pos, chunk_size, width), axis =1)\n",
    "print(chr_dataf.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Info 02:56:03] Reading GENCODE file (/Users/ds26/Projects/GenomePlotter/data/GENCODE.merged.bed.gz)...\n",
      "[Info 02:56:03] Selecting intersecting features.\n",
      "[Info 02:56:04] Assign intersecting features to each chunk.\n",
      "        chr   start     end  GC_ratio  x  y          GENCODE    color\n",
      "start                                                                \n",
      "0      22.0     0.0   450.0       NaN  0  0  heterochromatin  #FFC1C1\n",
      "450    22.0   450.0   900.0       NaN  1  0  heterochromatin  #FFC1C1\n",
      "900    22.0   900.0  1350.0       NaN  2  0  heterochromatin  #FFC1C1\n",
      "1350   22.0  1350.0  1800.0       NaN  3  0  heterochromatin  #FFC1C1\n",
      "1800   22.0  1800.0  2250.0       NaN  4  0  heterochromatin  #FFC1C1\n",
      "[Info 02:56:07] Assigning colors to each chunk.\n",
      "[Info 02:56:11] Apply darkness filter.\n",
      "        chr   start     end  GC_ratio  x  y          GENCODE    color\n",
      "start                                                                \n",
      "0      22.0     0.0   450.0       NaN  0  0  heterochromatin  #ffc6af\n",
      "450    22.0   450.0   900.0       NaN  1  0  heterochromatin  #ffc6af\n",
      "900    22.0   900.0  1350.0       NaN  2  0  heterochromatin  #ffc6af\n",
      "1350   22.0  1350.0  1800.0       NaN  3  0  heterochromatin  #ffc6af\n",
      "1800   22.0  1800.0  2250.0       NaN  4  0  heterochromatin  #ffc6af\n"
     ]
    }
   ],
   "source": [
    "# Reading GENCODE file:\n",
    "print(\"[Info %s] Reading GENCODE file (%s)...\" % (get_now(), GENCODE_file_loc))\n",
    "GENCODE_bed = pybedtools.BedTool(GENCODE_file_loc)\n",
    "\n",
    "# Run intersectbed:\n",
    "chr_data_bed = pybedtools.BedTool(dataFile)\n",
    "\n",
    "print(\"[Info %s] Selecting intersecting features.\" %(get_now()))\n",
    "GencodeIntersect = chr_data_bed.intersect(GENCODE_bed, wa = True, wb = True)\n",
    "GC_INT = pybedtools.bedtool.BedTool.to_dataframe(GencodeIntersect)\n",
    "\n",
    "# Assign features to each chunk:\n",
    "print(\"[Info %s] Assign intersecting features to each chunk.\" % (get_now()))\n",
    "GENCODE_chunks = GC_INT.groupby('start').apply(lambda x: 'exon' if 'exon' in x.thickEnd.unique() else 'gene' )\n",
    "GENCODE_chunks.names = \"GENCODE\"\n",
    "\n",
    "chr_dataf['GENCODE'] = 'intergenic' # By default, all GENCODE values are intergenic\n",
    "chr_dataf.GENCODE.update(GENCODE_chunks) # This value will be overwritten if overlaps with exon or gene\n",
    "chr_dataf.loc[chr_dataf.GC_ratio.isnull(), 'GENCODE'] = 'heterochromatin'\n",
    "print(chr_dataf.head())\n",
    "\n",
    "# Reading cytoband file:\n",
    "cytoband_file = '/Users/ds26/Projects/GenomePlotter/data/cytoBand.GRCh38.bed.bgz'\n",
    "cyb_df = pd.read_csv(cytoband_file, compression='gzip', sep='\\t')\n",
    "centromer_loc = cyb_df.loc[(cyb_df.chr == '22') & (cyb_df.type == 'acen'),['start', 'end']]\n",
    "centromer_loc = (int(centromer_loc.start.min()), \n",
    "                 int(centromer_loc.end.max()))\n",
    "\n",
    "# Updating centromere:\n",
    "chr_dataf.loc[(chr_dataf.end > centromer_loc[0]) & (chr_dataf.start < centromer_loc[1]) , 'GENCODE'] = 'heterochromatin'\n",
    "\n",
    "# Assigning colors:\n",
    "print(\"[Info %s] Assigning colors to each chunk.\" % (get_now()))\n",
    "colors_GENCODE = {\n",
    "    'heterochromatin' : linear_gradient('#ffc6af', finish_hex='#ffc6af', n=20), # \n",
    "    'intergenic': linear_gradient('#42b79a', n=20), # gray\n",
    "    'exon': linear_gradient('#CDCD00', n=20), # Purple\n",
    "    'gene': linear_gradient('#4278b7', n=20)} # Goldenrod\n",
    "chr_dataf['color'] = chr_dataf.apply(lambda x: colors_GENCODE[x['GENCODE']][int(x[3]*20)] if not np.isnan(x[3]) else colors_GENCODE['heterochromatin'][0], axis = 1)\n",
    "\n",
    "# Get the colors darker based on column number:\n",
    "print(\"[Info %s] Apply darkness filter.\" % (get_now()))\n",
    "chr_dataf['color'] = chr_dataf.apply(color_darkener, axis = 1, args=(width, start, threshold))\n",
    "print(chr_dataf.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Info 01:14:21] Opening GWAS file..\n",
      "[Info 01:14:21] Calculating plot coordinates for overlapping GWAS sigals.\n"
     ]
    }
   ],
   "source": [
    "print(\"[Info %s] Opening GWAS file..\" %(get_now()))\n",
    "GWAS_file = pybedtools.BedTool(GWAS_file_loc)\n",
    "full_GWAS = pybedtools.bedtool.BedTool.to_dataframe(GWAS_file)\n",
    "GWAS_df = full_GWAS[full_GWAS.chrom == str(chromosome)]\n",
    "print(\"[Info %s] Calculating plot coordinates for overlapping GWAS sigals.\" %(get_now()))\n",
    "GWAS_df = GWAS_df.apply(generate_xy, args = (min_pos, chunk_size, width), axis =1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import cairosvg\n",
    "\n",
    "\n",
    "class SVG_plot:\n",
    "    '''\n",
    "    This class contains all the methods and data to create the genome plot in svg format.\n",
    "    Once the process is done, use can choose to save the svg or render using cairosvg.\n",
    "    Good luck boy.\n",
    "    '''\n",
    "\n",
    "    def __init__(self, width, height, pixel, centr_start, centr_end):\n",
    "        self.width = width\n",
    "        self.height = height\n",
    "        self.pixel = pixel\n",
    "        self.centr_start = centr_start\n",
    "        self.centr_end = centr_end\n",
    "        self.plot =  '<svg width=\"%s\" height=\"%s\">\\n' % (width*pixel, height*pixel)\n",
    "\n",
    "    # Adding square:\n",
    "    def draw_chunk(self, row):\n",
    "        '''\n",
    "        This function expects a row of a dataframe in which there must be the\n",
    "        following three columns:\n",
    "            x - column number\n",
    "            y - row number\n",
    "            color - color of the field in hexadecimal code.\n",
    "        '''\n",
    "        x = row['x']\n",
    "        y = row['y']\n",
    "        color = row['color']\n",
    "        self.plot += ('<rect x=\"%s\" y=\"%s\" width=\"%s\" height=\"%s\" style=\"stroke-width:1;stroke:%s; fill: %s\" />\\n' %\n",
    "            (x*self.pixel, y*self.pixel, self.pixel , self.pixel, color, color))\n",
    "\n",
    "    # Adding dot:\n",
    "    def draw_GWAS(self, row):\n",
    "        x = row['x']\n",
    "        y = row['y']\n",
    "        self.plot += ('<circle cx=\"%s\" cy=\"%s\" r=\"2\" stroke=\"%s\" stroke-width=\"1\" fill=\"%s\" />\\n' %\n",
    "            (x*self.pixel, y*self.pixel, \"black\", \"black\"))\n",
    "\n",
    "    def mark_centromere(self):\n",
    "        # Calculate centromere dimensions:\n",
    "        centr = (\n",
    "            int(self.centr_start / self.width * self.pixel),\n",
    "            int(self.centr_end / self.width * self.pixel),\n",
    "            int((self.centr_end + self.centr_start) / (self.width * 2) * self.pixel),\n",
    "            int((self.centr_end - self.centr_start) / (self.width * 2) * self.pixel)\n",
    "        )\n",
    "\n",
    "        # Marking centromere on the left:\n",
    "        self.plot += ('<path d=\"M %s %s C %s %s, %s %s, %s %s C %s %s, %s %s, %s %s Z\" fill=\"white\"/>' %\n",
    "                 ( 0, centr[0],\n",
    "                   0, centr[2], \n",
    "                   centr[3], centr[2], \n",
    "                   centr[3]*2, centr[2],\n",
    "                   centr[3], centr[2], \n",
    "                   0, centr[2], \n",
    "                   0, centr[1]))\n",
    "        \n",
    "        # Marking centromoere on the right:\n",
    "        self.plot += ('<path d=\"M %s %s C %s %s, %s %s, %s %s C %s %s, %s %s, %s %s Z\" fill=\"white\"/>' %\n",
    "                 ( self.width * self.pixel, centr[0],\n",
    "                   self.width * self.pixel, centr[2], \n",
    "                   self.width * self.pixel - centr[3], centr[2], \n",
    "                   self.width * self.pixel - centr[3]*2, centr[2],\n",
    "                   self.width * self.pixel - centr[3], centr[2], \n",
    "                   self.width * self.pixel, centr[2], \n",
    "                   self.width * self.pixel, centr[1]))        \n",
    "        \n",
    "    # Adding legend?\n",
    "    # Adding centromere?\n",
    "    # Adding frame?\n",
    "    # Adding custom annotation.\n",
    "\n",
    "\n",
    "    # Close svg document:\n",
    "    def __close(self):\n",
    "        self.plot += '</svg>\\n'\n",
    "\n",
    "    # Save svg file:\n",
    "    def save_svg(self, file_name):\n",
    "        self.__close() # Closing the svg tag.\n",
    "        f = open(file_name, 'w')\n",
    "        f.write(self.plot)\n",
    "        f.close()\n",
    "\n",
    "    # Render png file:\n",
    "    def save_png(self, file_name):\n",
    "        cairosvg.svg2png(bytestring=self.plot,write_to=file_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Info 02:56:50] Drawing svg... might take a while to complete.\n",
      "[Info 03:00:32] Adding GWAS hits.\n",
      "[Info 03:00:36] Marking centromoere.\n",
      "[Info 03:00:36] Saving svg file..\n",
      "[Info 03:00:36] Saving png file..\n"
     ]
    }
   ],
   "source": [
    "pixel = 3\n",
    "outputFileName = (\"chr%s.w.test.%s\" %(chromosome, width))\n",
    "plot = SVG_plot(chr_dataf.x.max(), chr_dataf.y.max(), pixel, \n",
    "                int(centromer_loc[0]/chunk_size), int(centromer_loc[1]/chunk_size))\n",
    "\n",
    "print(\"[Info %s] Drawing svg... might take a while to complete.\" % (get_now()))\n",
    "chr_dataf.apply(plot.draw_chunk, axis = 1)\n",
    "\n",
    "print(\"[Info %s] Adding GWAS hits.\" %(get_now()))\n",
    "GWAS_df.apply(plot.draw_GWAS, axis = 1 )\n",
    "\n",
    "print(\"[Info %s] Marking centromoere.\" %(get_now()))\n",
    "plot.mark_centromere()\n",
    "\n",
    "print(\"[Info %s] Saving svg file..\" % (get_now()))\n",
    "plot.save_svg(outputFileName + \".svg\")\n",
    "\n",
    "print(\"[Info %s] Saving png file..\" % (get_now()))\n",
    "plot.save_png(outputFileName + \".png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<svg width=\"1820\" height=\"1820.0\">\n",
      "<polygon stroke=\"black\" fill=\"#A569BD\" stroke-width=\"2\" points=\"32.0,55.0 23.00688563992516,58.37796001393669 23.440491353343617,48.78115294937453 17.448835300684152,41.272039986063305 26.70993272936774,38.71884705062547 31.999999999999996,30.700000000000003 37.29006727063226,38.71884705062547 46.55116469931585,41.272039986063305 40.55950864665638,48.78115294937452 40.99311436007484,58.37796001393669\" />\n",
      "\n"
     ]
    }
   ],
   "source": [
    "class SVG_plot:\n",
    "    '''\n",
    "    This class contains all the methods and data to create the genome plot in svg format.\n",
    "    Once the process is done, use can choose to save the svg or render using cairosvg.\n",
    "\n",
    "    margins = array of four integers corresponding to the left, upper, right and bottom margin.\n",
    "    '''\n",
    "\n",
    "    def __init__(self, width, height, pixel, margins = [0,0,0,0]):\n",
    "        self.width = width\n",
    "        self.height = height\n",
    "        self.pixel = pixel\n",
    "        self.margins = margins\n",
    "\n",
    "        # This is a scaled star for plotting custom annotation:\n",
    "        self.star = [e for ts in zip([self.__rotate(0, pixel, math.radians(alpha*36)) for alpha in range(0,9,2)],\n",
    "                                    [self.__rotate(0, pixel, math.radians(alpha*36), 1.7) for alpha in range(1,10,2)],) for e in ts]\n",
    "\n",
    "        self.plot =  '<svg width=\"%s\" height=\"%s\">\\n' % (width*pixel + margins[0] + margins[2], height*pixel + margins[1] + margins[3])\n",
    "\n",
    "    def __adjust_coord(self, x, y):\n",
    "        ''' This simple function just sifts the coordinates by the margins and corrects for the pixel size.'''\n",
    "        return (int(x*self.pixel + self.margins[0]),\n",
    "                int(y*self.pixel + self.margins[1]))\n",
    "\n",
    "    def add_assoc(self, row):\n",
    "        ''' Adding a star to the polt and the rsID and trait... once later'''\n",
    "        (x,y) = self.__adjust_coord(row['x'], row['y'])\n",
    "\n",
    "        # Offsetting the star:\n",
    "        self.plot += ('<polygon stroke=\"black\" fill=\"#A569BD\" stroke-width=\"2\" points=\"%s\" />\\n' %\n",
    "                      \" \".join([\",\".join([str(x + a[0]),str(y+a[1])]) for a in self.star]))\n",
    "\n",
    "        # Adding annotation:\n",
    "        # to be implemented.... not yet there...\n",
    "\n",
    "\n",
    "    # Adding square:\n",
    "    def draw_chunk(self, row):\n",
    "        '''\n",
    "        This function expects a row of a dataframe in which there must be the\n",
    "        following three columns:\n",
    "            x - column number\n",
    "            y - row number\n",
    "            color - color of the field in hexadecimal code.\n",
    "        '''\n",
    "        (x,y) = self.__adjust_coord(row['x'],row['y'])\n",
    "        color = row['color']\n",
    "        self.plot += ('<rect x=\"%s\" y=\"%s\" width=\"%s\" height=\"%s\" style=\"stroke-width:1;stroke:%s; fill: %s\" />\\n' %\n",
    "            (x, y, self.pixel , self.pixel, color, color))\n",
    "\n",
    "    # Adding dot:\n",
    "    def draw_GWAS(self, row):\n",
    "        (x,y) = self.__adjust_coord(row['x'],row['y'])\n",
    "        self.plot += ('<circle cx=\"%s\" cy=\"%s\" r=\"%s\" stroke=\"%s\" stroke-width=\"1\" fill=\"%s\" />\\n' %\n",
    "            (x, y, self.pixel * 0.75, \"black\", \"black\"))\n",
    "\n",
    "    def mark_centromere(self, centr_start, centr_end):\n",
    "\n",
    "        # calculating the y coordinates of the centromeres based on the chunk count:\n",
    "        start_y = int(centr_start / self.width)\n",
    "        end_y = int(centr_end / self.width)\n",
    "\n",
    "        # Marking centromere on the left:\n",
    "        self.plot += ('<path d=\"M %s %s C %s %s, %s %s, %s %s C %s %s, %s %s, %s %s Z\" fill=\"white\"/>\\n' %\n",
    "                  ( self.__adjust_coord(0, start_y) +\n",
    "                    self.__adjust_coord(0, (start_y+end_y)/2) +\n",
    "                    self.__adjust_coord((end_y-start_y), (start_y+end_y)/2) +\n",
    "                    self.__adjust_coord((end_y-start_y)*2, (start_y+end_y)/2) +\n",
    "                    self.__adjust_coord((end_y-start_y), (start_y+end_y)/2) +\n",
    "                    self.__adjust_coord(0, (start_y+end_y)/2) +\n",
    "                    self.__adjust_coord(0, end_y )))\n",
    "\n",
    "        # Marking centromoere on the right:\n",
    "        self.plot += ('<path d=\"M %s %s C %s %s, %s %s, %s %s C %s %s, %s %s, %s %s Z\" fill=\"white\"/>\\n' %\n",
    "                  ( self.__adjust_coord(self.width+1, start_y) +\n",
    "                    self.__adjust_coord(self.width+1, (start_y+end_y)/2) +\n",
    "                    self.__adjust_coord(self.width+1-(end_y-start_y), (start_y+end_y)/2) +\n",
    "                    self.__adjust_coord(self.width+1-(end_y-start_y)*2, (start_y+end_y)/2) +\n",
    "                    self.__adjust_coord(self.width+1-(end_y-start_y), (start_y+end_y)/2) +\n",
    "                    self.__adjust_coord(self.width+1, (start_y+end_y)/2) +\n",
    "                    self.__adjust_coord(self.width+1, end_y)))\n",
    "\n",
    "    # Adding legend?\n",
    "    # Adding centromere? DONE\n",
    "    # Adding frame?\n",
    "    # Adding custom annotation.\n",
    "    # Adding cytobands.\n",
    "    # Rotating and scaling vecot\n",
    "    def __rotate(self, x, y, angle, scale=1):\n",
    "        return ((x * math.cos(angle) - y * math.sin(angle))*scale,\n",
    "                (x * math.sin(angle) + y * math.cos(angle))*scale)\n",
    "\n",
    "    # Close svg document:\n",
    "    def __close(self):\n",
    "        self.plot += '</svg>\\n'\n",
    "\n",
    "    # Save svg file:\n",
    "    def save_svg(self, file_name):\n",
    "        self.__close() # Closing the svg tag.\n",
    "        f = open(file_name, 'w')\n",
    "        f.write(self.plot)\n",
    "        f.close()\n",
    "\n",
    "    # Render png file:\n",
    "    def save_png(self, file_name):\n",
    "        cairosvg.svg2png(bytestring=self.plot,write_to=file_name)\n",
    "\n",
    "    # R\n",
    "plot = SVG_plot(200, 40000/200, 9, margins = [10, 10, 10, 10])\n",
    "row = { 'x' : 2.5,\n",
    "       'y' : 4.1 }\n",
    "plot.add_assoc(row)\n",
    "print(plot.plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'centromer_loc' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-16-94821e649c9c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m centromer_coord = (\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcentromer_loc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mchunk_size\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mwidth\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mpixel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m     \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcentromer_loc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mchunk_size\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mwidth\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mpixel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcentromer_loc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mcentromer_loc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mchunk_size\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mwidth\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mpixel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcentromer_loc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mcentromer_loc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mchunk_size\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mwidth\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mpixel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'centromer_loc' is not defined"
     ]
    }
   ],
   "source": [
    "centromer_coord = (\n",
    "    int(centromer_loc[0] / (chunk_size * width) * pixel),\n",
    "    int(centromer_loc[1] / (chunk_size * width) * pixel),\n",
    "    int((centromer_loc[1] + centromer_loc[0]) / (chunk_size * width * 2) * pixel),\n",
    "    int((centromer_loc[1] - centromer_loc[0]) / (chunk_size * width * 2) * pixel)\n",
    ")\n",
    "\n",
    "# Drawing centromere:\n",
    "centr_svg = ('<path d=\"M %s %s C %s %s, %s %s, %s %s C %s %s, %s %s, %s %s Z\" fill=\"white\"/>\\n' %\n",
    "         ( 0, centromer_coord[0],\n",
    "           0, centromer_coord[2], centromer_coord[3], centromer_coord[2], centromer_coord[3]*2, centromer_coord[2],\n",
    "           centromer_coord[3], centromer_coord[2], 0, centromer_coord[2], 0, centromer_coord[1]\n",
    "           , ))\n",
    "centr_svg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<svg width=\"2000\" height=\"2000.0\">\n",
      "<path d=\"M 100 1189 C 100 1189.0, 100 1189.0, 100 1189.0 C 100 1189.0, 100 1189.0, 100 1189 Z\" fill=\"white\"/>\n",
      "<path d=\"M 1900 1189 C 1900 1189.0, 1900 1189.0, 1900 1189.0 C 1900 1189.0, 1900 1189.0, 1900 1189 Z\" fill=\"white\"/>\n",
      "<path d=\"M 100 1189 C 100 1189.0, 100 1189.0, 100 1189.0 C 100 1189.0, 100 1189.0, 100 1189 Z\" fill=\"white\"/>\n",
      "<path d=\"M 1900 1189 C 1900 1189.0, 1900 1189.0, 1900 1189.0 C 1900 1189.0, 1900 1189.0, 1900 1189 Z\" fill=\"white\"/>\n",
      "\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1800.0"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "40000/200*9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import math\n",
    "\n",
    "def __rotate(x,y,angle, scale=1):\n",
    "    return ((x * math.cos(angle) - y * math.sin(angle))*scale,\n",
    "            (x * math.sin(angle) + y * math.cos(angle))*scale)\n",
    "pixel = 1\n",
    "star = [e for ts in zip([rotate(0, pixel, alpha*36,2) for alpha in range(1,10,2)], [rotate(0, pixel, alpha*36) for alpha in range(0,9,2)]) for e in ts]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'13.98355770688623,19.74407262074519 12.0,21.0 10.14636298916443,20.751019195534024 11.746176637237964,19.03274941172612 13.60230527146766,18.803079861884285 12.49102159389847,20.871147401032342 10.753975577992694,21.564424219884543 11.303941511655088,19.282014916030285 12.808130438912722,18.17053964412925 12.85550437075082,20.517795588650813'"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = 12\n",
    "y = 20\n",
    "\" \".join([\",\".join([str(x + a[0]),str(y+a[1])]) for a in star])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Raw Cell Format",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
